{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2e9aeba0-1800-4065-991b-0676770caea6",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Background"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4bb275d-0e96-4f34-bb5a-cd2eb040cbda",
   "metadata": {},
   "source": [
    "This RNA-seq workflow is adapted from the RNA-seq tutuorial at [rnabio.org](https://rnabio.org).\n",
    "\n",
    "``` Malachi Griffith*, Jason R. Walker, Nicholas C. Spies, Benjamin J. Ainscough, Obi L. Griffith*. 2015. Informatics for RNA-seq: A web resource for analysis on the cloud. PLoS Comp Biol. 11(8):e1004393```\n",
    "\n",
    "The [ENCODE project](https://www.encodeproject.org) publishes [guidelines and best practices for RNA-seq](https://www.encodeproject.org/documents/cede0cbe-d324-4ce7-ace4-f0c3eddf5972/@@download/attachment/ENCODE%20Best%20Practices%20for%20RNA_v2.pdf) that are a must read."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99c94222-8085-4976-b73a-b5c5a2a11eaa",
   "metadata": {},
   "source": [
    "## Initial setup\n",
    "The tutorial being followed uses AWS cloud but I am writing this pipeline to be ran locally on a linux-based machine. More specifically, I am using Manjaro with kernel version 6.1.7-1-MANJARO (64-bit) and KDE plasma version 5.26.4 on X11.\n",
    "\n",
    "[Configure your environment](https://rnabio.org/module-00-setup/0000/09/01/Environment/) and then [install all necessary software](https://rnabio.org/module-00-setup/0000/10/01/Installation/) according to the rnabio website. The specific steps you follow while doing this will depend on your system and how you want to install the tools. I would suggest looking for maintained packages for your specific distribution before trying to install from source.\n",
    "\n",
    "Specific verison information of software installed on my setup:\n",
    "- SAMtools v1.16.1, installed from AUR\n",
    "    - run with ```samtools``` from CL\n",
    "- bam-readcount v1.0.1, compiled from source\n",
    "    - run from ```~$HOME/bioinformatics/bam-readcount/build/bin``` with ```./bam-readcount```\n",
    "- HISAT2 v2.2.1-1, installed from AUR\n",
    "    - run with ```hisat2``` from CL\n",
    "- stringtie v2.2.1, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/stringtie-2.2.1.Linux_x86_64``` with ```./stringtie```\n",
    "- gffcompare v0.12.6-2, installed from AUR\n",
    "    - run with ```gffcompare``` from CL\n",
    "- htseq-count v.0.11.3-1, installed from AUR and also in bioinformatics conda env\n",
    "    - run by importing in Python?\n",
    "- tophat v2.1.1, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/tophat-2.1.1.Linux_x86_64``` with ```./gtf_to_fasta```\n",
    "- kallisto v0.48.0-1, installed from AUR\n",
    "    - run with ```kallisto``` from CL\n",
    "- fastqc v0.11.9-1, installed from AUR\n",
    "    - run with ```fastqc``` from CL\n",
    "- multiqc v1.14, installed with pip3 in bioinformatics conda env\n",
    "    - run with ```multiqc --help``` from CL\n",
    "- picard v2.26.4, downloaded as JAR file\n",
    "    - run from ```~$HOME/bioinformatics/``` with ```java -jar picard.jar```\n",
    "- flexbar v3.5.0, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/flexbar-3.5.0-linux``` with ```./flexbar```\n",
    "- regtools v1.0.0, installed from AUR (git source)\n",
    "    - run with ```regtools``` from CL\n",
    "- rseqc v5.0.1, installed with pip3 in bioinformatics conda env\n",
    "    - import in python with ```rseqc``` or run with ```read_GC.py``` in CL?\n",
    "- bedops v2.4.40, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/bedops_linux_x86_64-v2.4.40/bin``` with```./bedops``` and ```./gff2bed```\n",
    "- gtfToGenePred v unknown, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/gtfToGenePred``` with ```./gtfToGenePred```\n",
    "- genePredToBed v unknown, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/genePredToBed``` with ```./genePredToBed```\n",
    "- how_are_we_stranded_here, v unknown, installed with pip3 in bioinformatics conda env\n",
    "    - run with ```check_strandedness``` in CL\n",
    "- cell ranger v7.1.0, have to register to download, pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/cellranger-7.1.0``` with ```./cellranger```\n",
    "- tabix, website only offers apt installation, not relevant on arch-based distro. Tabix seems to be included with HTSlib which is already installed as a dependency for samtools and bcftools\n",
    "- BWA v0.7.17=r1188, compiled from source\n",
    "    - run from ```~$HOME/bioinformatics/bwa``` with ```./bwa``` or ```bwa```\n",
    "- BCFtools v1.16, installed from AUR\n",
    "    - run with ```bcftools``` from CL\n",
    "- peddy v0.4.8, cloned and installed with pip in bioinformatics conda env\n",
    "    - run with ```peddy``` from CL in conda env\n",
    "- slivar v0.2.7, pre-complied binary\n",
    "    - run from ```~$HOME/bioinformatics``` with ```./slivar```\n",
    "- STRling v0.5.1, pre-complied binary\n",
    "    - run from ```~$HOME/bioinformatics``` with ```./strling```\n",
    "- freebayes v1.3.6, pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics``` with ```./freebayes-1.3.6-linux-amd64-static```\n",
    "\n",
    "### R Libraries needed:\n",
    "- devtools\n",
    "- dplyr\n",
    "- gplots\n",
    "- ggplot2\n",
    "\n",
    "install with \n",
    "```install.packages(c(\"devtools\",\"dplyr\",\"gplots\",\"ggplot2\"),repos=\"http://cran.us.r-project.org\")```\n",
    "\n",
    "### Bioconductor libraries needed:\n",
    "- [genefilter](http://bioconductor.org/packages/release/bioc/html/genefilter.html)\n",
    "- [ballgown](http://bioconductor.org/packages/release/bioc/html/ballgown.html)\n",
    "- [edgeR](http://www.bioconductor.org/packages/release/bioc/html/edgeR.html)\n",
    "- [GenomicRanges](http://bioconductor.org/packages/release/bioc/html/GenomicRanges.html)\n",
    "- [rhdf5](https://www.bioconductor.org/packages/release/bioc/html/rhdf5.html)\n",
    "- [biomaRt](https://bioconductor.org/packages/release/bioc/html/biomaRt.html)\n",
    "\n",
    "### Sleuth\n",
    "[Source](https://pachterlab.github.io/sleuth/download)\n",
    "\n",
    "devtools need to be installed if not already ```install.packages(\"devtools\")```\n",
    "install with ```devtools::install_github(\"pachterlab/sleuth\")```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1afe64c2-2756-44ba-9b0b-e9189f2d2313",
   "metadata": {},
   "source": [
    "# General notes on RNA-seq workflow\n",
    "\n",
    "Due to inconsistencies in the naming conventions for chromosomes, you should obtain your reference genome and annotation packages from the same source (Ensembl, NCBI, UCSC). Additionally, your annotations must correspond with the same reference genome version (i.e. both correspond to NCBI human GRCh38).\n",
    "\n",
    "[This guide](https://pmbio.org/module-02-inputs/0002/02/01/Reference_Genome/) explains reference genomes and how to obtain them.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff56a121-6803-4e44-9c4a-0fafa432329d",
   "metadata": {},
   "source": [
    "# Obtain known gene/transcript annotations\n",
    "\n",
    "Download the annotations from Ensembl via ```wget http://genomedata.org/rnaseq-tutorial/annotations/GRCh38/chr22_with_ERCC92.gtf```. \n",
    "\n",
    "You can preview some of the file with ```cat chr22_with_ERCC92.gtf | column -t | less -p exon -S```.\n",
    "\n",
    "\n",
    "```cat chr22_with_ERCC92.gtf | grep -w gene | wc -l```\n",
    "\n",
    "You can view the structure of one transcript with ```grep ENST00000342247 $RNA_REF_GTF | less -p \"exon\\s\" -S```. Exit with q.\n",
    "\n",
    "Other sources of annontation files for a HISAT2/StringTie/Ballgown pipeline, refer to [here](https://rnabio.org/module-01-inputs/0001/03/01/Annotations/).\n",
    "\n",
    "HISAT2 has prebuilt indexes for DNA and RNA for both Ensembl and UCSC genomes [here](https://ccb.jhu.edu/software/hisat2/index.shtml).\n",
    "\n",
    "**Important**: chromosome names in the .gtf file must be the same as those in the reference genome .fa file. If StringTie returns expression values that are all 0, this is the likley cause of that error. Download the annotation and reference genome files from the same source.\n",
    "\n",
    "**Make sure your reference genome and annotations are for the same build of the genome (i.e. both for GRCh38).**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f10eeb8-390b-4706-b4b4-6650f0ce9940",
   "metadata": {},
   "source": [
    "# Creating index with HISAT2\n",
    "\n",
    "This was done directly using python files from the [HISAT2 github repo](https://github.com/DaehwanKimLab/hisat2), not sure how to do it using the installed program listed above or if it's possible. Just clone the repo and use the files necessary.\n",
    "\n",
    "Run the following commands from the location of your reference .gtf and .fa files. The chr22_with_ERCC92.gtf and chr22_with_ERCC92.fa. You may need a ```./``` in front of the first two commands\n",
    "\n",
    "```hisat2_extract_splice_sites.py chr22_with_ERCC92.gtf > splicesites.tsv```\n",
    "\n",
    "```hisat2_extract_exons.py chr22_with_ERCC92.gtf >exons.tsv```\n",
    "\n",
    "```hisat2-build -p 8 --ss splicesites.tsv --exon exons.tsv chr22_with_ERCC92.fa chr22_with_ERCC92```\n",
    "\n",
    "The output should be a collection of binary files with names like chr22_with_ERCC92.x.ht2, the x being a number."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b06fc2fd-5eff-424d-b4a7-296c410a9ce9",
   "metadata": {},
   "source": [
    "# RNA-seq test data\n",
    "\n",
    "```mkdir test_data && cd test_data```\n",
    "\n",
    "```wget http://genomedata.org/rnaseq-tutorial/HBR_UHR_ERCC_ds_5pc.tar```\n",
    "\n",
    "unpack the tarball\n",
    "\n",
    "```tar -xvf HBR_UHR_ERCC_ds_5pc.tar```\n",
    "\n",
    "Look at the head of one of the files.\n",
    "\n",
    "```zcat UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz | head -n 8```\n",
    "\n",
    "Determine how many reads are in the first library.\n",
    "\n",
    "```zcat UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz | grep -P \"^\\@HWI\" | wc -l```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "796edb8a-4e09-4ba4-812c-8d06e5bc8df3",
   "metadata": {},
   "source": [
    "# Check_strandedness (optional, but not really)\n",
    "\n",
    "This step is not required but due to differences in usage of RNA-seq library construction kits, it is **highly** recommended to empirically verify strandedness using the below method.\n",
    "\n",
    "The check_strandedness should already be installed at the beginning. It can also be installed as a [Docker image](https://hub.docker.com/r/mgibio/checkstrandedness).\n",
    "\n",
    "Convert gtf to genepred\n",
    "\n",
    "```gtfToGenePred chr22_with_ERCC92.gtf chr22_with_ERCC92.genePred```\n",
    "\n",
    "Convert genPred to bed12\n",
    "\n",
    "```genePredToBed chr22_with_ERCC92.genePred chr22_with_ERCC92.bed12```\n",
    "\n",
    "Use bedtools to create fasta from GTF\n",
    "\n",
    "```bedtools getfasta -fi chr22_with_ERCC92.fa -bed chr22_with_ERCC92.bed12 -s -split -name -fo chr22_ERCC92.transcripts.fa```\n",
    "\n",
    "Clean up the header lines of the transcript file\n",
    "\n",
    "```cat chr22_ERCC92_transcripts.fa | perl -ne 'if($_ =~/^\\>\\S+\\:\\:(ERCC\\-\\d+)\\:.*/){print \">$1\\n\"}elsif ($_ =~/^\\>(\\S+)\\:\\:.*/){print \">$1\\n\"}else{print $_}' > chr22_ERCC92_transcripts.clean.fa```\n",
    "\n",
    "View the clean file\n",
    "\n",
    "```less chr22_ERCC92_transcripts.clean.fa```\n",
    "\n",
    "Add in transcript_id field to .gtf file\n",
    "\n",
    "```awk '{ if ($0 ~ \"transcript_id\") print $0; else print $0\" transcript_id \\\"\\\";\"; }' chr22_with_ERCC92.gtf > chr22_with_ERCC92_tidy.gtf```\n",
    "\n",
    "check strandedness \n",
    "\n",
    "```check_strandedness --gtf chr22_with_ERCC92_tidy.gtf --transcripts chr22_ERCC92_transcripts.clean.fa --reads_1 test_data/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz --reads_2 test_data/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz```\n",
    "\n",
    "OUTPUT:\n",
    "\n",
    "Fraction of reads failed to determine: 0.1123\n",
    "\n",
    "Fraction of reads explained by \"1++,1--,2+-,2-+\": 0.0155 (1.7% of explainable reads)\n",
    "\n",
    "Fraction of reads explained by \"1+-,1-+,2++,2--\": 0.8722 (98.3% of explainable reads)\n",
    "\n",
    "Over 90% of reads explained by \"1+-,1-+,2++,2--\"\n",
    "\n",
    "Data is likely RF/fr-firststrand\n",
    "\n",
    "More info on strandedness can be found [here](https://rnabio.org/module-09-appendix/0009/12/01/StrandSettings/)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "764cd5a0-fa7a-4dc1-9f67-335e69350306",
   "metadata": {},
   "source": [
    "# Pre-alignment QC\n",
    "\n",
    "To get a sense of the raw reads quality in your fastq files, run the following programs.\n",
    "\n",
    "```fastqc *.fastq.gz``` in the same directory as your reads.\n",
    "\n",
    "It will output an HTML file for each read in the same directory that contains a full report of the quality measures.\n",
    "\n",
    "In the \"over-represented sequences\" section, try [BLASTing](https://blast.ncbi.nlm.nih.gov/Blast.cgi) the sequences to try determining a source organism.\n",
    "\n",
    "Run multiqc to combine all the fastqc reports.\n",
    "\n",
    "```multiqc .``` run in the same directory as the reports.\n",
    "\n",
    "Make a new directory for the fastqc .html outputs and move them\n",
    "\n",
    "```mkdir fastqc && mv *_fastqc* fastqc```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b9099f7-50d6-42da-91bc-2853d7b6a3bf",
   "metadata": {},
   "source": [
    "# Trim adapters\n",
    "\n",
    "Make a new folder for the trimmed reads\n",
    "\n",
    "```mkdir trimmed```\n",
    "\n",
    "Download the Illumina adapter sequence files, or the relevant files for your sequencing machine.\n",
    "\n",
    "```wget http://genomedata.org/rnaseq-tutorial/illumina_multiplex.fa```\n",
    "\n",
    "Use flexbar to remove adapter sequences and trim first 13 bases of each read. Adjust the number of threads to something manageable for your system. It is set to 24 threads and takes a few seconds to run for each command. It may take much longer with a slower CPU.\n",
    "\n",
    "Possible options for flexbar and their meaning:\n",
    "\n",
    "   - ```–adapter-min-overlap 7``` requires a minimum of 7 bases to match the adapter\n",
    "   - ```–adapter-trim-end RIGHT``` uses a trimming strategy to remove the adapter from the 3 prime or RIGHT end of the read\n",
    "   - ```–max-uncalled 300``` allows as many as 300 uncalled or N bases (MiSeq read lengths can be 300bp)\n",
    "   - ```–min-read-length``` the minimum read length allowed after trimming is 25bp.\n",
    "   - ```–threads 8``` use 8 threads\n",
    "   - ```–zip-output GZ``` the input FASTQ files are gzipped so we will output gzipped FASTQ to save space\n",
    "   - ```–adapters``` define the path to the adapter FASTA file to trim\n",
    "   - ```–reads``` define the path to the read 1 FASTQ file of reads\n",
    "   - ```–reads2``` define the path to the read 2 FASTQ file of reads\n",
    "   - ```–target``` a base path for the output files. The value will _1.fastq.gz and _2.fastq.gz for read 1 and read 2 respectively\n",
    "   - ```–pre-trim-left``` trim a fixed number of bases at left read end. For example, to trim 5 bases at the left side of reads: –pre-trim-left 5\n",
    "   - ```–pre-trim-right``` trim a fixed number of bases at right read end. For example, to trim 5 bases at the right side of reads: –pre-trim-right 5\n",
    "   - ```–pre-trim-phred``` trim based on phred quality value to deal with higher error rates towards the end of reads. For example, to trim the 3 prime end until quality offset value 30 or higher is reached, specify: –pre-trim-phred 30\n",
    "\n",
    "The commands to run for flexbar for this dataset are as follows:\n",
    "\n",
    "```./flexbar --adapter-min-overlap 7 --adapter-trim-end RIGHT --adapters /home/user/bioinformatics/ref_genome/illumina_multiplex.fa --pre-trim-left 13 --max-uncalled 300 --min-read-length 25 --threads 24 --zip-output GZ --reads /home/user/bioinformatics/ref_genome/test_data/UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz --reads2 /home/user/bioinformatics/ref_genome/test_data/UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz --target /home/user/bioinformatics/ref_genome/test_data/trimmed/UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22```\n",
    "\n",
    "```./flexbar --adapter-min-overlap 7 --adapter-trim-end RIGHT --adapters /home/user/bioinformatics/ref_genome/illumina_multiplex.fa --pre-trim-left 13 --max-uncalled 300 --min-read-length 25 --threads 24 --zip-output GZ --reads /home/user/bioinformatics/ref_genome/test_data/UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz --reads2 /home/user/bioinformatics/ref_genome/test_data/UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz --target /home/user/bioinformatics/ref_genome/test_data/trimmed/UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22```\n",
    "\n",
    "```./flexbar --adapter-min-overlap 7 --adapter-trim-end RIGHT --adapters /home/user/bioinformatics/ref_genome/illumina_multiplex.fa --pre-trim-left 13 --max-uncalled 300 --min-read-length 25 --threads 24 --zip-output GZ --reads /home/user/bioinformatics/ref_genome/test_data/UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz --reads2 /home/user/bioinformatics/ref_genome/test_data/UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz --target /home/user/bioinformatics/ref_genome/test_data/trimmed/UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22```\n",
    "\n",
    "```./flexbar --adapter-min-overlap 7 --adapter-trim-end RIGHT --adapters /home/user/bioinformatics/ref_genome/illumina_multiplex.fa --pre-trim-left 13 --max-uncalled 300 --min-read-length 25 --threads 24 --zip-output GZ --reads /home/user/bioinformatics/ref_genome/test_data/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz --reads2 /home/user/bioinformatics/ref_genome/test_data/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz --target /home/user/bioinformatics/ref_genome/test_data/trimmed/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22```\n",
    "\n",
    "```./flexbar --adapter-min-overlap 7 --adapter-trim-end RIGHT --adapters /home/user/bioinformatics/ref_genome/illumina_multiplex.fa --pre-trim-left 13 --max-uncalled 300 --min-read-length 25 --threads 24 --zip-output GZ --reads /home/user/bioinformatics/ref_genome/test_data/HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz --reads2 /home/user/bioinformatics/ref_genome/test_data/HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz --target /home/user/bioinformatics/ref_genome/test_data/trimmed/HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22```\n",
    "\n",
    "```./flexbar --adapter-min-overlap 7 --adapter-trim-end RIGHT --adapters /home/user/bioinformatics/ref_genome/illumina_multiplex.fa --pre-trim-left 13 --max-uncalled 300 --min-read-length 25 --threads 24 --zip-output GZ --reads /home/user/bioinformatics/ref_genome/test_data/HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz --reads2 /home/user/bioinformatics/ref_genome/test_data/HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz --target /home/user/bioinformatics/ref_genome/test_data/trimmed/HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22```\n",
    "\n",
    "Compare the impact of trimming with the original QC reports by running fastqc and multiqc on the trimmed reads\n",
    "\n",
    "In the directory of the trimmed reads, run:\n",
    "\n",
    "```fastqc *.fastq.gz```\n",
    "\n",
    "```multiqc .```\n",
    "\n",
    "Same as before, make a new directory for the fastqc .html outputs and move them\n",
    "\n",
    "```mkdir fastqc && mv *_fastqc* fastqc```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf703eb9-5218-4e71-80c4-8b4e376b6eff",
   "metadata": {},
   "source": [
    "# Alignment with HISAT2\n",
    "\n",
    "For more details on HISAT2, refer to the [manual on Github](http://daehwankimlab.github.io/hisat2/manual/)\n",
    "\n",
    "The outputs after running HISAT2 will be a SAM/BAM file for each data set.\n",
    "\n",
    "Options for HISAT2 include:\n",
    "\n",
    "\n",
    "   - ```-p 8``` tells HISAT2 to use eight CPUs for bowtie alignments.\n",
    "   - ```–rna-strandness RF``` specifies strandness of RNAseq library. We will specify RF since the TruSeq strand-specific library was used to make these libraries. See here for options.\n",
    "   - ```–rg-id $ID``` specifies a read group ID that is a unique identifier.\n",
    "   - ```–rg SM:$SAMPLE_NAME``` specifies a read group sample name. This together with rg-id will allow you to determine which reads came from which sample in the merged bam later on.\n",
    "   - ```–rg LB:$LIBRARY_NAME``` specifies a read group library name. This together with rg-id will allow you to determine which reads came from which library in the merged bam later on.\n",
    "   - ```–rg PL:ILLUMINA``` specifies a read group sequencing platform.\n",
    "   - ```–rg PU:$PLATFORM_UNIT``` specifies a read group sequencing platform unit. Typically this consists of FLOWCELL-BARCODE.LANE\n",
    "   - ```–dta``` Reports alignments tailored for transcript assemblers.\n",
    "   - ```-x /path/to/hisat2/index``` The HISAT2 index filename prefix (minus the trailing .X.ht2) built earlier including splice sites and exons.\n",
    "   - ```-1 /path/to/read1.fastq.gz``` The read 1 FASTQ file, optionally gzip(.gz) or bzip2(.bz2) compressed.\n",
    "   - ```-2 /path/to/read2.fastq.gz``` The read 2 FASTQ file, optionally gzip(.gz) or bzip2(.bz2) compressed.\n",
    "   - ```-S /path/to/output.sam``` The output SAM format text file of alignments.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6a505ea-9119-403b-a24a-2483f82ede2f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
