{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "994d63a5-4f40-4b33-b26d-7368d1cc4b57",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "This RNA-seq pipeline uses the reference genome GRCm39 and HISAT2 for alignment and HTseq-count for generating raw counts. The rest of the analysis, at the moment, is done using iDEP.96"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0511186b-9542-467b-beb4-542afe1e04cb",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Initial setup\n",
    "The tutorial being followed uses AWS cloud but I am writing this pipeline to be ran locally on a linux-based machine. More specifically, I am using Manjaro with kernel version 6.1.11-1-MANJARO (64-bit).\n",
    "\n",
    "[Configure your environment](https://rnabio.org/module-00-setup/0000/09/01/Environment/) and then [install all necessary software](https://rnabio.org/module-00-setup/0000/10/01/Installation/) according to the rnabio website. The specific steps you follow while doing this will depend on your system and how you want to install the tools. I would suggest looking for maintained packages for your specific distribution before trying to install from source.\n",
    "\n",
    "Specific verison information of software installed on my setup:\n",
    "- SAMtools v1.16.1, installed from AUR\n",
    "    - run with ```samtools``` from CL\n",
    "- bam-readcount v1.0.1, compiled from source\n",
    "    - run from ```~$HOME/bioinformatics/bam-readcount/build/bin``` with ```./bam-readcount```\n",
    "- HISAT2 v2.2.1-1, installed from AUR\n",
    "    - run with ```hisat2``` from CL\n",
    "- stringtie v2.2.1, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/stringtie-2.2.1.Linux_x86_64``` with ```./stringtie```\n",
    "- gffcompare v0.12.6-2, installed from AUR\n",
    "    - run with ```gffcompare``` from CL\n",
    "- htseq-count v.0.11.3-1, installed from AUR and also in bioinformatics conda env\n",
    "    - run with ```htseq-count``` from terminal with conda bioinformatics env active\n",
    "- tophat v2.1.1, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/tophat-2.1.1.Linux_x86_64``` with ```./gtf_to_fasta```\n",
    "- kallisto v0.48.0-1, installed from AUR\n",
    "    - run with ```kallisto``` from CL\n",
    "- fastqc v0.11.9-1, installed from AUR\n",
    "    - run with ```fastqc``` from CL\n",
    "- multiqc v1.14, installed with pip3 in bioinformatics conda env\n",
    "    - run with ```multiqc --help``` from CL\n",
    "- picard v2.26.4, downloaded as JAR file\n",
    "    - run from ```~$HOME/bioinformatics/``` with ```java -jar picard.jar```\n",
    "- flexbar v3.5.0, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/flexbar-3.5.0-linux``` with ```./flexbar```\n",
    "- regtools v1.0.0, installed from AUR (git source)\n",
    "    - run with ```regtools``` from CL\n",
    "- rseqc v5.0.1, installed with pip3 in bioinformatics conda env\n",
    "    - import in python with ```rseqc``` or run with ```read_GC.py``` in CL?\n",
    "- bedops v2.4.40, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/bedops_linux_x86_64-v2.4.40/bin``` with```./bedops``` and ```./gff2bed```\n",
    "- gtfToGenePred v unknown, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/gtfToGenePred``` with ```./gtfToGenePred```\n",
    "- genePredToBed v unknown, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/genePredToBed``` with ```./genePredToBed```\n",
    "- how_are_we_stranded_here, v unknown, installed with pip3 in bioinformatics conda env\n",
    "    - run with ```check_strandedness``` in CL\n",
    "- cell ranger v7.1.0, have to register to download, pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/cellranger-7.1.0``` with ```./cellranger```\n",
    "- tabix, website only offers apt installation, not relevant on arch-based distro. Tabix seems to be included with HTSlib which is already installed as a dependency for samtools and bcftools\n",
    "- BWA v0.7.17=r1188, compiled from source\n",
    "    - run from ```~$HOME/bioinformatics/bwa``` with ```./bwa``` or ```bwa```\n",
    "- BCFtools v1.16, installed from AUR\n",
    "    - run with ```bcftools``` from CL\n",
    "- peddy v0.4.8, cloned and installed with pip in bioinformatics conda env\n",
    "    - run with ```peddy``` from CL in conda env\n",
    "- slivar v0.2.7, pre-complied binary\n",
    "    - run from ```~$HOME/bioinformatics``` with ```./slivar```\n",
    "- STRling v0.5.1, pre-complied binary\n",
    "    - run from ```~$HOME/bioinformatics``` with ```./strling```\n",
    "- freebayes v1.3.6, pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics``` with ```./freebayes-1.3.6-linux-amd64-static```\n",
    "\n",
    "### R Libraries needed:\n",
    "- devtools\n",
    "- dplyr\n",
    "- gplots\n",
    "- ggplot2\n",
    "- UpSetR\n",
    "\n",
    "install with \n",
    "```install.packages(c(\"devtools\",\"dplyr\",\"gplots\",\"ggplot2\",\"UpSetR\"),repos=\"http://cran.us.r-project.org\")```\n",
    "\n",
    "### Bioconductor libraries needed:\n",
    "- [genefilter](http://bioconductor.org/packages/release/bioc/html/genefilter.html)\n",
    "- [ballgown](http://bioconductor.org/packages/release/bioc/html/ballgown.html)\n",
    "- [edgeR](http://www.bioconductor.org/packages/release/bioc/html/edgeR.html)\n",
    "- [GenomicRanges](http://bioconductor.org/packages/release/bioc/html/GenomicRanges.html)\n",
    "- [rhdf5](https://www.bioconductor.org/packages/release/bioc/html/rhdf5.html)\n",
    "- [biomaRt](https://bioconductor.org/packages/release/bioc/html/biomaRt.html)\n",
    "- [sva](https://www.bioconductor.org/packages/release/bioc/html/sva.html)\n",
    "- [GAGE](https://bioconductor.org/packages/release/bioc/html/gage.html)\n",
    "\n",
    "### Sleuth\n",
    "[Source](https://pachterlab.github.io/sleuth/download)\n",
    "\n",
    "devtools need to be installed if not already ```install.packages(\"devtools\")```\n",
    "\n",
    "install with ```devtools::install_github(\"pachterlab/sleuth\")```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74d17e40-40b0-4860-9b0a-83787751ca87",
   "metadata": {},
   "source": [
    "***\n",
    "# General notes on RNA-seq workflow\n",
    "\n",
    "Due to inconsistencies in the naming conventions for chromosomes, you should obtain your reference genome and annotation packages from the same source (Ensembl, NCBI, UCSC). Additionally, your annotations must correspond with the same reference genome version (i.e. both correspond to NCBI human GRCh38).\n",
    "\n",
    "[This guide](https://pmbio.org/module-02-inputs/0002/02/01/Reference_Genome/) explains reference genomes and how to obtain them under the \"Reference Genome Options\" section at the bottom.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f040157b-f241-4195-a36c-097b7f86df3d",
   "metadata": {},
   "source": [
    "***\n",
    "# Obtain reference genome and known gene/transcript annotations\n",
    "\n",
    "### Reference genome\n",
    "\n",
    "The Mus musculus GRCm39 reference genome in FASTA format was downloaded ```wget https://ftp.ensembl.org/pub/release-109/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.primary_assembly.fa.gz```\n",
    "\n",
    "Location: ```/home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.dna.primary_assembly.fa```\n",
    "\n",
    "### Gene annotations in .gtf format\n",
    "\n",
    "Gene annotations in .gtf format were downloaded from Ensembl as well via the command ```wget ftp://ftp.ensembl.org/pub/release-109/gtf/mus_musculus/Mus_musculus.GRCm39.109.gtf.gz```.\n",
    "\n",
    "Location: ```/home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.gtf```\n",
    "\n",
    "You can preview the file with ```head -n 50 /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.gtf | column -t | less -p exon -S```\n",
    "\n",
    "```cat /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.gtf | grep -w gene | wc -l```\n",
    "\n",
    "You can view the structure of one transcript with ```grep ENSMUSG00000102693 /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.gtf | less -p \"exon\\s\" -S```\n",
    "\n",
    "**Important**: chromosome names in the .gtf file must be the same as those in the reference genome .fa file. If StringTie returns expression values that are all 0, this is the likley cause of that error. Download the annotation and reference genome files from the same source.\n",
    "\n",
    "**Make sure your reference genome and annotations are for the same build of the genome (i.e. both for GRCh38).**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "582e596f-2d4f-49c5-9841-ad0958963545",
   "metadata": {},
   "source": [
    "***\n",
    "# Creating index with HISAT2\n",
    "\n",
    "**WARNING** This step will take **A LOT** of RAM, well over 100 gb so make sure the system has enough.\n",
    "\n",
    "It is much easier and faster, if possible, to download a pre-built index provided by the HISAT2 devs at [their website](https://daehwankimlab.github.io/hisat2/download/).\n",
    "\n",
    "I downloaded the GRCm38 index, files were renamed with ```rename genome Mus_musculus.GRCm38.dna.primary_assembly *.ht2```.\n",
    "\n",
    "This was done directly using python files from the [HISAT2 github repo](https://github.com/DaehwanKimLab/hisat2), not sure how to do it using the installed program listed above or if it's possible. Just clone the repo and use the files necessary.\n",
    "\n",
    "Switch to hisat cloned repo ```cd /home/user/bioinformatics/hisat2```, or provide a full file path.\n",
    "\n",
    "Run the following commands.\n",
    "\n",
    "```/home/user/bioinformatics/hisat2/hisat2_extract_splice_sites.py /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.gtf > /home/user/bioinformatics/reference_genomes/GRCm39/splicesites.tsv```\n",
    "\n",
    "```/home/user/bioinformatics/hisat2/hisat2_extract_exons.py /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.gtf > /home/user/bioinformatics/reference_genomes/GRCm39/exons.tsv```\n",
    "\n",
    "I have hisat2 installed separately from the repo, hence why I don't need to provide a file path. This index was not built on my machine, it was done on a Google Cloud Platform instance with 256 GB of RAM.\n",
    "\n",
    "```hisat2-build -p 24 --ss /home/user/bioinformatics/reference_genomes/GRCm39/splicesites.tsv --exon /home/user/bioinformatics/reference_genomes/GRCm39/exons.tsv /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.dna.primary_assembly.fa /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.dna.primary_assembly```\n",
    "\n",
    "The output should be a collection of binary files with names like Mus_musculus.GRCm38.dna.primary_assembly.x.ht2 or .rf, the x being a number."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11d2c932-2621-462d-892c-c71cc7372b0f",
   "metadata": {},
   "source": [
    "***\n",
    "# Look at some of the actual RNA-seq data\n",
    "\n",
    "Look at the head of the files.\n",
    "\n",
    "```cd ~/bioinformatics/PyMT_Exercise_Paper_2023```\n",
    "\n",
    "```zcat EP_01_S1_R1_001.fastq.gz | head -n 100```\n",
    "\n",
    "A cursory look shows us that the average read length appears to be around 50 bp. Since these are paired-end reads, that makes sense. For the quality lines, most seem to be ```C```, which corresponds to a Phred quality score of 33, or a 5% chance that the base was incorrectly called.\n",
    "\n",
    "Determine how many reads are in a file.\n",
    "\n",
    "```zcat EP_01_S1_R1_001.fastq.gz | grep -P \"^\\@VH00\" | wc -l```\n",
    "\n",
    "Note, you may have to change the grep command to match the instrument ID.\n",
    "\n",
    "There is a bash file in the Github repo under the PyMT_Exercise_Paper_2023 folder titled ```total_PE_read_counts.sh``` that will count all the reads for all FASTQ files and output them to a txt file. Simply define the directory for the FASTQ files, the output directory for the text file, and if there is a different naming scheme besides ```R1``` and ```R2``` for the paired reads, change those values in the script. You may also need to change the search value for grep from ```VH00``` to what is in your FASTQ files.\n",
    "\n",
    "### After looking at the total read counts there are some potential issues\n",
    "\n",
    "This script confirms all read pairs are equal in length."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c5becb9-95ab-452b-8c8f-0ee038c30435",
   "metadata": {},
   "source": [
    "***\n",
    "# Check strandedness\n",
    "\n",
    "This step is not required but due to differences in usage of RNA-seq library construction kits, it is **highly** recommended to empirically verify strandedness using the below method.\n",
    "\n",
    "The check_strandedness should already be installed at the beginning. It can also be installed as a [Docker image](https://hub.docker.com/r/mgibio/checkstrandedness).\n",
    "\n",
    "I suppose putting these in the GRCm39 folder should work for now.\n",
    "\n",
    "```/home/user/bioinformatics/reference_genomes/GRCm39```\n",
    "\n",
    "Convert gtf to genepred\n",
    "\n",
    "```cd /home/user/bioinformatics/gtfToGenePred```, or just provide full file path.\n",
    "\n",
    "```/home/user/bioinformatics/gtfToGenePred/gtfToGenePred /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.gtf /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.genePred```\n",
    "\n",
    "Convert genePred to bed12\n",
    "\n",
    "```cd /home/user/bioinformatics/genePredToBed```, or just provide full file path.\n",
    "\n",
    "```/home/user/bioinformatics/genePredToBed/genePredToBed /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.genePred /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.bed12```\n",
    "\n",
    "Use bedtools to create fasta index from GTF\n",
    "\n",
    "```bedtools getfasta -fi /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.dna.primary_assembly.fa -bed /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.bed12 -s -split -name -fo /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.dna.primary_assembly.transcripts.fa```\n",
    "\n",
    "Clean up the header lines of the transcript file\n",
    "\n",
    "```cat /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.dna.primary_assembly.transcripts.fa | perl -ne 'if($_ =~/^\\>\\S+\\:\\:(ERCC\\-\\d+)\\:.*/){print \">$1\\n\"}elsif ($_ =~/^\\>(\\S+)\\:\\:.*/){print \">$1\\n\"}else{print $_}' > /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.dna.primary_assembly.transcripts.clean.fa```\n",
    "\n",
    "View the clean file\n",
    "\n",
    "```less /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.dna.primary_assembly.transcripts.clean.fa```\n",
    "\n",
    "Add in transcript_id field to .gtf file\n",
    "\n",
    "```awk '{ if ($0 ~ \"transcript_id\") print $0; else print $0\" transcript_id \\\"\\\";\"; }' /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.gtf > /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109_tidy.gtf```\n",
    "\n",
    "check strandedness\n",
    "\n",
    "**Everything up to here should be done already**.\n",
    "\n",
    "```conda activate bioinformatics```\n",
    "\n",
    "This command saves some files to the directory it is run from, best to change to where the fastq files are.\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Exercise_Paper_2023```\n",
    "\n",
    "```check_strandedness --gtf /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109_tidy.gtf --transcripts /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.dna.primary_assembly.transcripts.clean.fa --reads_1 /home/user/bioinformatics/PyMT_Exercise_Paper_2023/EP_01_S1_R1_001.fastq.gz --reads_2 /home/user/bioinformatics/PyMT_Exercise_Paper_2023/EP_01_S1_R2_001.fastq.gz```\n",
    "\n",
    "OUTPUT:\n",
    "\n",
    "Total 200000 usable reads were sampled\n",
    "\n",
    "This is PairEnd Data\n",
    "\n",
    "Fraction of reads failed to determine: 0.1196\n",
    "\n",
    "Fraction of reads explained by \"1++,1--,2+-,2-+\": 0.0044 (0.5% of explainable reads)\n",
    "\n",
    "Fraction of reads explained by \"1+-,1-+,2++,2--\": 0.8760 (99.5% of explainable reads)\n",
    "\n",
    "Over 90% of reads explained by \"1+-,1-+,2++,2--\"\n",
    "\n",
    "Data is likely RF/fr-firststrand\n",
    "\n",
    "More info on strandedness can be found [here](https://rnabio.org/module-09-appendix/0009/12/01/StrandSettings/)\n",
    "\n",
    "Based on the strandedness, this sequencing was likely done using the dUTP method."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "305030d2-d505-454c-a5bf-ae6608fe5d99",
   "metadata": {},
   "source": [
    "***\n",
    "# Pre-alignment QC\n",
    "\n",
    "To get a sense of the raw reads quality in your fastq files, run the following programs.\n",
    "\n",
    "### fastqc [docs](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/)\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Exercise_Paper_2023```\n",
    "\n",
    "```fastqc -t 24 *.fastq.gz``` in the same directory as your reads.\n",
    "\n",
    "It will output an HTML file for each read in the same directory that contains a full report of the quality measures.\n",
    "\n",
    "In the \"over-represented sequences\" section, try [BLASTing](https://blast.ncbi.nlm.nih.gov/Blast.cgi) the sequences to try determining a source organism.\n",
    "\n",
    "### multiqc\n",
    "\n",
    "Run multiqc to combine all the fastqc reports.\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Exercise_Paper_2023```\n",
    "\n",
    "Run ```multiqc .``` in the same directory as the reports.\n",
    "\n",
    "Renamed file to ```pre-alignment_multiqc_report.html```\n",
    "\n",
    "Make a new directory for the fastqc .html outputs and move them\n",
    "\n",
    "```mkdir fastqc && mv *_fastqc* fastqc```\n",
    "\n",
    "## Bash script\n",
    "\n",
    "I made a bash script to handle this step, ```pre-alignment_QC.sh```. All that should need to be changed is the fastq_dir variable. Which should point to where the FASTQ files are. The script does not yet account for any sub-folders.\n",
    "\n",
    "## View reports\n",
    "\n",
    "Looking at the MultiQC report, none of the warnings are concerning or unusual. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74b597a8-7053-43b2-9bc2-885281ec1894",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "***\n",
    "# Alignment with HISAT2\n",
    "\n",
    "For more details on HISAT2, refer to the [manual on Github](http://daehwankimlab.github.io/hisat2/manual/)\n",
    "\n",
    "The outputs after running HISAT2 will be a SAM/BAM file for each data set.\n",
    "\n",
    "Options for HISAT2 include:\n",
    "\n",
    "\n",
    "   - ```-p 24``` tells HISAT2 to use 24 threads for bowtie alignments.\n",
    "   - ```–rna-strandness``` specifies strandness of RNAseq library. Check_strandedness said it is likely RF/fr. So we remove use R/RF.\n",
    "   - ```–rg-id $ID``` specifies a read group ID that is a unique identifier.\n",
    "   - ```–rg SM:$SAMPLE_NAME``` specifies a read group sample name. This together with rg-id will allow you to determine which reads came from which sample in the merged bam later on.\n",
    "   - ```–rg LB:$LIBRARY_NAME``` specifies a read group library name. This together with rg-id will allow you to determine which reads came from which library in the merged bam later on.\n",
    "   - ```–rg PL:ILLUMINA``` specifies a read group sequencing platform.\n",
    "   - ```–rg PU:$PLATFORM_UNIT``` specifies a read group sequencing platform unit. Typically this consists of FLOWCELL-BARCODE.LANE\n",
    "   - ```–dta``` Reports alignments tailored for transcript assemblers.\n",
    "   - ```-x /path/to/hisat2/index``` The HISAT2 index filename prefix (minus the trailing .X.ht2) built earlier including splice sites and exons.\n",
    "   - ```-1 /path/to/read1.fastq.gz``` The read 1 FASTQ file, optionally gzip(.gz) or bzip2(.bz2) compressed.\n",
    "   - ```-2 /path/to/read2.fastq.gz``` The read 2 FASTQ file, optionally gzip(.gz) or bzip2(.bz2) compressed.\n",
    "   - ```-U /path/to/unstranded.fastq.gz``` If data is unstranded, provide link to FASTQ or .gz version here.\n",
    "   - ```-S /path/to/output.sam``` The output SAM format text file of alignments.\n",
    "   \n",
    "Create an alignment folder for the outputs.\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Exercise_Paper_2023 && mkdir -p alignments && cd alignments```\n",
    "\n",
    "Run a command for each of the replicates. I will try running based off the trimmed fastq files and see how it goes.\n",
    "\n",
    "I wrote a script to handle this, ```HISAT2_alignment.sh```\n",
    "\n",
    "```hisat2 -p 24 --rg-id EP_01_S1_001 --rg SM:EP_01_S1_001 --rg LB:EP_01_S1_001_Lib --rg PL:ILLUMINA -x /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.dna.primary_assembly --rna-strandness RF --dta -1 /home/user/bioinformatics/PyMT_Exercise_Paper_2023/EP_01_S1_001.read1.fastq.gz -2 /home/user/bioinformatics/PyMT_Exercise_Paper_2023/EP_01_S1_001.read2.fastq.gz -S /home/user/bioinformatics/PyMT_Exercise_Paper_2023/EP_01_S1_001.sam```\n",
    "\n",
    "\n",
    "### Convert SAM to BAM files\n",
    "\n",
    "I also wrote a script for this, ```Convert_SAM_to_BAM.sh```\n",
    "\n",
    "```-@ int``` is the number of threads to use.\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments```\n",
    "\n",
    "```samtools sort -@ 24 -o EP_01_S1_001.bam EP_01_S1_001.sam```\n",
    "\n",
    "And so on so forth...\n",
    "\n",
    "The SAM files are very large and will take up hundreds of gigs, so run ```rm -f *.sam``` in the alignments directory.\n",
    "\n",
    "### Merge all BAM files\n",
    "Other tools such as \"samtools merge\" and \"bamtools merge\" can also achieve this. Create a BAM file for each group. Can't do this yet because I'm not sure which sample belongs to which group.\n",
    "\n",
    "```java -Xmx2g -jar /home/user/bioinformatics/picard.jar MergeSamFiles -OUTPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/FC.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P11B_FC1_S1_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P11D_FC2_S2_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P13B_FC3_S3_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P13C_FC4_S4_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P13E_FC5_S5_L001_001.bam```\n",
    "\n",
    "```java -Xmx2g -jar /home/user/bioinformatics/picard.jar MergeSamFiles -OUTPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/FT.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P12BL_FT1_S6_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P12CRR_FT2_S7_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P13A_FT3_S8_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P14E_FT4_S9_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P13F_FT5_S10_L001_001.bam```\n",
    "\n",
    "```java -Xmx2g -jar /home/user/bioinformatics/picard.jar MergeSamFiles -OUTPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/MC.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P186-5_MC1_S11_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P186-7_MC2_S12_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P189-1_MC3_S13_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P187-1_MC4_S14_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P188-2_MC5_S15_L001_001.bam```\n",
    "\n",
    "```java -Xmx2g -jar /home/user/bioinformatics/picard.jar MergeSamFiles -OUTPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/MT.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P187-3_MT1_S16_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P186-6_MT2_S17_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P188-1_MT3_S18_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P188-4_MT4_S19_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P188-5_MT5_S20_L001_001.bam```\n",
    "\n",
    "You can verify that all bam files were created successfully, there should be 24.\n",
    "\n",
    "In the same directory as the aligned .bam files, ```ls -l *.bam | wc -l```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "868c581b-09b0-4ea6-8ca4-24cf0be5ed81",
   "metadata": {},
   "source": [
    "***\n",
    "# Using the Integrative Genomics Viewer (IGV)\n",
    "\n",
    "```cd /home/user/bioinformatics/IGV_Linux_2.15.4```\n",
    "\n",
    "IGV can be run from the download directory using ```./igv.sh```\n",
    "\n",
    "To re-visit the shortish tutorial on navigating IGV, go [here](https://rnabio.org/module-02-alignment/0002/04/01/IGV/). You can follow the tutorial there using the provided files.\n",
    "\n",
    "For our dataset, we need to index the BAM files before loading them into IGV. This will be done with samtools. From the directory where the BAM files are located, run:\n",
    "\n",
    "You can just run the following command to index all .bam files in the directory. This way is easier than running multiple commands.\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments```\n",
    "\n",
    "```find *.bam -exec echo samtools index -@28 {} \\; | sh```\n",
    "\n",
    "Load the .bam files in IGV and look around. The .bam.bai index needs to be in the same directory as the .bam files."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b31f311f-bd5e-4410-83b6-1ee1ee71d00d",
   "metadata": {},
   "source": [
    "***\n",
    "# Alignment QC\n",
    "\n",
    "The command ```samtools flagstat``` can provide a summary of the alignments.\n",
    "\n",
    "```mkdir -p /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments/flagstat```\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments```\n",
    "\n",
    "```samtools flagstat /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments/EP_01_S1_001.bam > /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments/EP_01_S1_001.bam.flagstat```\n",
    "\n",
    "And so on....\n",
    "\n",
    "Or, you can also use ```find``` to do it in one command.\n",
    "\n",
    "```find *_*.bam -exec echo samtools flagstat {} \\> flagstat/{}.flagstat \\; | sh```\n",
    "\n",
    "### FastQc on bam file\n",
    "\n",
    "From your alignment directory (where the bam files are).\n",
    "\n",
    "```mkdir /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments/fastqc```\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments```\n",
    "\n",
    "```fastqc -t 24 *.bam```\n",
    "\n",
    "Clean up the current directory and move the reports into the fastqc folder.\n",
    "\n",
    "```mv *fastqc.html fastqc/ && mv *fastqc.zip fastqc/```\n",
    "\n",
    "### Picard\n",
    "\n",
    "Using Picard we can generate metrics and figures that express the quality of our RNA-seq data.\n",
    "\n",
    "We first need to create input files for Picard based on the reference genome so it can properly run the function CollectRnaSeqMetrics.\n",
    "\n",
    "```java -jar /home/user/bioinformatics/picard.jar CreateSequenceDictionary -R /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.dna.primary_assembly.fa -O /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.dna.primary_assembly.dict```\n",
    "\n",
    "Create a bed file of the location of ribosomal sequences in our reference (first extract from the gtf then convert to bed). Note that here we pull all the \"rrna\" transcripts from the GTF.\n",
    "\n",
    "```grep --color=none -i -P \"rrna\" /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.gtf > /home/user/bioinformatics/reference_genomes/GRCm39/ref_ribosome.gtf```\n",
    "\n",
    "Note: ```gff2bed``` is part of BEDOPS. Run the script directly from the bedops/bin folder. You may need to add the /bin folder to your path variable ```export PATH=\"$PATH:/home/user/bioinformatics/bedops_linux_x86_64-v2.4.40/bin\"```, otherwise convert2bed may not be found.\n",
    "\n",
    "```cd /home/user/bioinformatics/bedops_linux_x86_64-v2.4.40/bin```\n",
    "\n",
    "```gff2bed < /home/user/bioinformatics/reference_genomes/GRCm39/ref_ribosome.gtf > /home/user/bioinformatics/reference_genomes/GRCm39/ref_ribosome.bed```\n",
    "\n",
    "Create interval list file to identify the location of the ribosomal sequences in the reference\n",
    "\n",
    "```java -jar /home/user/bioinformatics/picard.jar BedToIntervalList -I /home/user/bioinformatics/reference_genomes/GRCm39/ref_ribosome.bed -O /home/user/bioinformatics/reference_genomes/GRCm39/ref_ribosome.interval_list -SD /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.dna.primary_assembly.dict```\n",
    "\n",
    "Create genePred file for the reference transcriptome\n",
    "\n",
    "```cd /home/user/bioinformatics/gtfToGenePred```\n",
    "\n",
    "```./gtfToGenePred -genePredExt /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.gtf /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.ref_flat.txt```\n",
    "\n",
    "Reformat the genePred file\n",
    "\n",
    "```cat /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.ref_flat.txt | awk '{print $12\"\\t\"$0}' | cut -d$'\\t' -f1-11 > /home/user/bioinformatics/reference_genomes/GRCm39/tmp.txt```\n",
    "\n",
    "```mv /home/user/bioinformatics/reference_genomes/GRCm39/tmp.txt /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.ref_flat.txt```\n",
    "\n",
    "Go to the directory with your alignments.\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments && mkdir picard```\n",
    "\n",
    "This data is stranded paired-end reads, specifically ```RF/fr-firststrand stranded (dUTP)``` as determined in the check strandedness step, so the STRAND flag should be set to ```STRAND=SECOND_READ_TRANSCRIPTION_STRAND```.\n",
    "\n",
    "```find *_*.bam -exec echo java -jar /home/user/bioinformatics/picard.jar CollectRnaSeqMetrics I={} O=picard/{}.RNA_Metrics REF_FLAT=/home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.ref_flat.txt STRAND=SECOND_READ_TRANSCRIPTION_STRAND RIBOSOMAL_INTERVALS=/home/user/bioinformatics/reference_genomes/GRCm39/ref_ribosome.interval_list \\; | sh```\n",
    "\n",
    "### RSeQC\n",
    "\n",
    "Another tool for generating RNA-seq QC reports.\n",
    "\n",
    "You will need\n",
    "- aligned bam files\n",
    "- index file for each bam file\n",
    "- transcript bed file (in bed12 format)\n",
    "\n",
    "Start by converting the reference gtf to genePred. This step has probably already been done.\n",
    "\n",
    "```gtfToGenePred /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.gtf /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.genePred```\n",
    "\n",
    "Then convert the genePred to bed12. This step also probably done already.\n",
    "\n",
    "```genePredToBed /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.genePred /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.bed12```\n",
    "\n",
    "Go back to the alignment directory. Make sure the bioinformatics env is active.\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments && mkdir rseqc```\n",
    "\n",
    "```geneBody_coverage.py -i Pistilli_P11B_FC1_S1_L001_001.bam, Pistilli_P11D_FC2_S2_L001_001.bam, Pistilli_P13B_FC3_S3_L001_001.bam, Pistilli_P13C_FC4_S4_L001_001.bam, Pistilli_P13E_FC5_S5_L001_001.bam -r /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.bed12 -o rseqc/FC```\n",
    "\n",
    "```geneBody_coverage.py -i Pistilli_P12BL_FT1_S6_L001_001.bam, Pistilli_P12CRR_FT2_S7_L001_001.bam, Pistilli_P13A_FT3_S8_L001_001.bam, Pistilli_P14E_FT4_S9_L001_001.bam, Pistilli_P13F_FT5_S10_L001_001.bam -r /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.bed12 -o rseqc/FT```\n",
    "\n",
    "```geneBody_coverage.py -i Pistilli_P186-5_MC1_S11_L001_001.bam, Pistilli_P186-7_MC2_S12_L001_001.bam, Pistilli_P189-1_MC3_S13_L001_001.bam, Pistilli_P187-1_MC4_S14_L001_001.bam, Pistilli_P188-2_MC5_S15_L001_001.bam -r /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.bed12 -o rseqc/MC```\n",
    "\n",
    "```geneBody_coverage.py -i Pistilli_P187-3_MT1_S16_L001_001.bam, Pistilli_P186-6_MT2_S17_L001_001.bam, Pistilli_P188-1_MT3_S18_L001_001.bam, Pistilli_P188-4_MT4_S19_L001_001.bam, Pistilli_P188-5_MT5_S20_L001_001.bam -r /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.bed12 -o rseqc/MT```\n",
    "\n",
    "And then that same geneBody_coverage.py command for each group.\n",
    "\n",
    "The following commands should also be run from the alignment directory ```cd /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments```\n",
    "\n",
    "```find *_*.bam -exec echo inner_distance.py -i {} -r /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.bed12 -o rseqc/{} \\; | sh && find *_*.bam -exec echo junction_annotation.py -i {} -r /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.bed12 -o rseqc/{} \\; | sh && find *_*.bam -exec echo junction_saturation.py -i {} -r /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.bed12 -o rseqc/{} \\; | sh && find *_*.bam -exec echo read_distribution.py -i {} -r /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.bed12 \\> rseqc/{}.read_dist.txt \\; | sh && find *_*.bam -exec echo RNA_fragment_size.py -i {} -r /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.bed12 \\> rseqc/{}.frag_size.txt \\; | sh && find *_*.bam -exec echo bam_stat.py -i {} \\> {}.bam_stat.txt \\; | sh && rm -f log.txt```\n",
    "\n",
    "### MultiQC\n",
    "\n",
    "Now we will compile a multiQC report from all of the QC done above.\n",
    "\n",
    "In the alignment directory ```cd /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments```.\n",
    "\n",
    "```multiqc ./```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eab94d3a-47d6-435d-a4d6-19687fe7d5b8",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "***\n",
    "# Expression Analysis\n",
    "\n",
    "For this run of the pipeline, **I will not be generating normalized expression values and instead only the raw counts, as below**.\n",
    "\n",
    "### BamQC\n",
    "\n",
    "Before generating counts, especially raw counts, it is useful to perform some QC on the aligned BAM files produced by HISAT2. This will let us better examine the quality of our alignments and set a reasonable quality filter value in the ```--minaqual``` flag for HTSeq-count.\n",
    "\n",
    "This is done using the [BamQC](https://github.com/s-andrews/BamQC) program. After compiling the program, you can generate a report for your BAM file using the following command:\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments```\n",
    "\n",
    "Ran the below command for now. You can also run this on BAM \n",
    "\n",
    "```/home/user/bioinformatics/BamQC/bin/bamqc --gff /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.gtf *.bam```\n",
    "\n",
    "You can list multiple BAM files in one command. It will save the report to the same directory as your BAM file.\n",
    "\n",
    "Of particular interest is the \"Mapping Quality Distribution\" section, as it will show the spread of the MAPQ quality values generated by HISAT2. Which should range from 0 to 60. A value closer to 60 means a higher quality alignment.\n",
    "\n",
    "Looking at the BamQC files for each group, the majority of reads, around 50-60% are MAPQ score of 60, the rest are 1 or 0. Based on running the HTSeq commands with 1, 10, and 40, the counts do not really change. If there was a more varied spread, then 40 would probably be a good value, but here it doesn't change anything since there aren't any alignments with a score between 2 and 59.\n",
    "\n",
    "### Generate raw counts instead of FPKM/TPM values\n",
    "\n",
    "If you only want to generate the raw counts this can be done using HTSeq-count [docs](https://htseq.readthedocs.io/).\n",
    "\n",
    "Some notes about the flags used:\n",
    "\n",
    "\n",
    "- ```–format``` specify the input file format one of BAM or SAM. Since we have BAM format files, select ‘bam’ for this option.\n",
    "- ```–order``` provide the expected sort order of the input file. Previously we generated position sorted BAM files so use ‘pos’.\n",
    "- ```–mode``` determines how to deal with reads that overlap more than one feature. Various sources say ‘intersection-strict’ mode is best in most cases.\n",
    "- ```–stranded``` specifies whether data is stranded or not. The TruSeq strand-specific RNA libraries suggest the ‘reverse’ option for this parameter. This data is RF/fr-firststrand so the flag is set to 'reverse'.\n",
    "- ```–minaqual``` will skip all reads with alignment quality lower than the given minimum value.\n",
    "- ```–type``` specifies the feature type (3rd column in GFF file) to be used. (default, suitable for RNA-Seq and Ensembl GTF files: exon)\n",
    "- ```–idattr``` The feature ID used to identify the counts in the output table. The default, suitable for RNA-SEq and Ensembl GTF files, is gene_id.\n",
    "\n",
    "Change directories to where the expression folder is ```cd /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments```\n",
    "\n",
    "```mkdir -p expression/htseq_counts && cd expression/htseq_counts```\n",
    "\n",
    "Run each htseq-count command to get the raw counts for each read file. So the command below for each .bam file. Or, use the \"run_htseq-count.sh\" script to do them all in parallel because htseq-count can't multithread.\n",
    "\n",
    "```htseq-count --format bam --order pos --mode intersection-strict --stranded reverse --minaqual 1 --type exon --idattr gene_id /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments/EP_01_S1_001.bam /home/user/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.109.gtf > /home/user/bioinformatics/PyMT_Exercise_Paper_2023/alignments/expression/htseq_counts/EP_01_gene.tsv\"\"\n",
    "\n",
    "Merge all the results files into one file.\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Paper_2023/alignments/expression/htseq_counts```\n",
    "\n",
    "```join FC1_gene.tsv FC2_gene.tsv | join - FC3_gene.tsv | join - FC4_gene.tsv | join - FC5_gene.tsv | join - FT1_gene.tsv | join - FT2_gene.tsv | join - FT3_gene.tsv | join - FT4_gene.tsv | join - FT5_gene.tsv | join - MC1_gene.tsv | join - MC2_gene.tsv | join - MC3_gene.tsv | join - MC4_gene.tsv | join - MC5_gene.tsv | join - MT1_gene.tsv | join - MT2_gene.tsv | join - MT3_gene.tsv | join - MT4_gene.tsv | join - MT5_gene.tsv > gene_read_counts_table_all.tsv```\n",
    "\n",
    "```echo \"GeneID FC1 FC2 FC3 FC4 FC5 FT1 FT2 FT3 FT4 FT5 MC1 MC2 MC3 MC4 MC5 MT1 MT2 MT3 MT4 MT5\" > header.txt```\n",
    "\n",
    "```cat header.txt gene_read_counts_table_all.tsv | grep -v \"__\" | awk -v OFS=\"\\t\" '$1=$1' > gene_read_counts_table_all_final.tsv```\n",
    "\n",
    "If you're curious, ```grep -v \"__\"``` is used to filter out summary lines at the end of htseq-count files that did not align, had poor quality alignment, etc.\n",
    "\n",
    "```rm -f gene_read_counts_table_all.tsv header.txt```\n",
    "\n",
    "```head gene_read_counts_table_all_final.tsv | column -t```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "445456e1-d058-4780-97ea-685a95845bc8",
   "metadata": {
    "tags": []
   },
   "source": [
    "***\n",
    "# Differential Expression (DESeq2)\n",
    "\n",
    "The differential expression is done using the bioconductor package [DESeq2](https://master.bioconductor.org/packages/release/bioc/html/DESeq2.html).\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Paper_2023/alignments```\n",
    "\n",
    "```mkdir -p de/deseq2/ && cd de/deseq2/```\n",
    "\n",
    "Start an R session by typing ```R``` in a terminal.\n",
    "\n",
    "The following code is an R script that will perform DE using Ballgown. You can run each command line by line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5111fb72-f306-4cd7-bdee-55f501545fc7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
