{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "994d63a5-4f40-4b33-b26d-7368d1cc4b57",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "This RNA-seq pipeline uses HISAT2 for alignment and HTseq-count for generating raw counts. The rest of the analysis, at the moment, is done using iDEP.96"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0511186b-9542-467b-beb4-542afe1e04cb",
   "metadata": {},
   "source": [
    "## Initial setup\n",
    "The tutorial being followed uses AWS cloud but I am writing this pipeline to be ran locally on a linux-based machine. More specifically, I am using Manjaro with kernel version 6.1.11-1-MANJARO (64-bit).\n",
    "\n",
    "[Configure your environment](https://rnabio.org/module-00-setup/0000/09/01/Environment/) and then [install all necessary software](https://rnabio.org/module-00-setup/0000/10/01/Installation/) according to the rnabio website. The specific steps you follow while doing this will depend on your system and how you want to install the tools. I would suggest looking for maintained packages for your specific distribution before trying to install from source.\n",
    "\n",
    "Specific verison information of software installed on my setup:\n",
    "- SAMtools v1.16.1, installed from AUR\n",
    "    - run with ```samtools``` from CL\n",
    "- bam-readcount v1.0.1, compiled from source\n",
    "    - run from ```~$HOME/bioinformatics/bam-readcount/build/bin``` with ```./bam-readcount```\n",
    "- HISAT2 v2.2.1-1, installed from AUR\n",
    "    - run with ```hisat2``` from CL\n",
    "- stringtie v2.2.1, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/stringtie-2.2.1.Linux_x86_64``` with ```./stringtie```\n",
    "- gffcompare v0.12.6-2, installed from AUR\n",
    "    - run with ```gffcompare``` from CL\n",
    "- htseq-count v.0.11.3-1, installed from AUR and also in bioinformatics conda env\n",
    "    - run with ```htseq-count``` from terminal with conda bioinformatics env active\n",
    "- tophat v2.1.1, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/tophat-2.1.1.Linux_x86_64``` with ```./gtf_to_fasta```\n",
    "- kallisto v0.48.0-1, installed from AUR\n",
    "    - run with ```kallisto``` from CL\n",
    "- fastqc v0.11.9-1, installed from AUR\n",
    "    - run with ```fastqc``` from CL\n",
    "- multiqc v1.14, installed with pip3 in bioinformatics conda env\n",
    "    - run with ```multiqc --help``` from CL\n",
    "- picard v2.26.4, downloaded as JAR file\n",
    "    - run from ```~$HOME/bioinformatics/``` with ```java -jar picard.jar```\n",
    "- flexbar v3.5.0, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/flexbar-3.5.0-linux``` with ```./flexbar```\n",
    "- regtools v1.0.0, installed from AUR (git source)\n",
    "    - run with ```regtools``` from CL\n",
    "- rseqc v5.0.1, installed with pip3 in bioinformatics conda env\n",
    "    - import in python with ```rseqc``` or run with ```read_GC.py``` in CL?\n",
    "- bedops v2.4.40, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/bedops_linux_x86_64-v2.4.40/bin``` with```./bedops``` and ```./gff2bed```\n",
    "- gtfToGenePred v unknown, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/gtfToGenePred``` with ```./gtfToGenePred```\n",
    "- genePredToBed v unknown, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/genePredToBed``` with ```./genePredToBed```\n",
    "- how_are_we_stranded_here, v unknown, installed with pip3 in bioinformatics conda env\n",
    "    - run with ```check_strandedness``` in CL\n",
    "- cell ranger v7.1.0, have to register to download, pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/cellranger-7.1.0``` with ```./cellranger```\n",
    "- tabix, website only offers apt installation, not relevant on arch-based distro. Tabix seems to be included with HTSlib which is already installed as a dependency for samtools and bcftools\n",
    "- BWA v0.7.17=r1188, compiled from source\n",
    "    - run from ```~$HOME/bioinformatics/bwa``` with ```./bwa``` or ```bwa```\n",
    "- BCFtools v1.16, installed from AUR\n",
    "    - run with ```bcftools``` from CL\n",
    "- peddy v0.4.8, cloned and installed with pip in bioinformatics conda env\n",
    "    - run with ```peddy``` from CL in conda env\n",
    "- slivar v0.2.7, pre-complied binary\n",
    "    - run from ```~$HOME/bioinformatics``` with ```./slivar```\n",
    "- STRling v0.5.1, pre-complied binary\n",
    "    - run from ```~$HOME/bioinformatics``` with ```./strling```\n",
    "- freebayes v1.3.6, pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics``` with ```./freebayes-1.3.6-linux-amd64-static```\n",
    "\n",
    "### R Libraries needed:\n",
    "- devtools\n",
    "- dplyr\n",
    "- gplots\n",
    "- ggplot2\n",
    "- UpSetR\n",
    "\n",
    "install with \n",
    "```install.packages(c(\"devtools\",\"dplyr\",\"gplots\",\"ggplot2\",\"UpSetR\"),repos=\"http://cran.us.r-project.org\")```\n",
    "\n",
    "### Bioconductor libraries needed:\n",
    "- [genefilter](http://bioconductor.org/packages/release/bioc/html/genefilter.html)\n",
    "- [ballgown](http://bioconductor.org/packages/release/bioc/html/ballgown.html)\n",
    "- [edgeR](http://www.bioconductor.org/packages/release/bioc/html/edgeR.html)\n",
    "- [GenomicRanges](http://bioconductor.org/packages/release/bioc/html/GenomicRanges.html)\n",
    "- [rhdf5](https://www.bioconductor.org/packages/release/bioc/html/rhdf5.html)\n",
    "- [biomaRt](https://bioconductor.org/packages/release/bioc/html/biomaRt.html)\n",
    "- [sva](https://www.bioconductor.org/packages/release/bioc/html/sva.html)\n",
    "- [GAGE](https://bioconductor.org/packages/release/bioc/html/gage.html)\n",
    "\n",
    "### Sleuth\n",
    "[Source](https://pachterlab.github.io/sleuth/download)\n",
    "\n",
    "devtools need to be installed if not already ```install.packages(\"devtools\")```\n",
    "\n",
    "install with ```devtools::install_github(\"pachterlab/sleuth\")```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74d17e40-40b0-4860-9b0a-83787751ca87",
   "metadata": {},
   "source": [
    "***\n",
    "# General notes on RNA-seq workflow\n",
    "\n",
    "Due to inconsistencies in the naming conventions for chromosomes, you should obtain your reference genome and annotation packages from the same source (Ensembl, NCBI, UCSC). Additionally, your annotations must correspond with the same reference genome version (i.e. both correspond to NCBI human GRCh38).\n",
    "\n",
    "[This guide](https://pmbio.org/module-02-inputs/0002/02/01/Reference_Genome/) explains reference genomes and how to obtain them under the \"Reference Genome Options\" section at the bottom.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f040157b-f241-4195-a36c-097b7f86df3d",
   "metadata": {},
   "source": [
    "***\n",
    "# Obtain reference genome and known gene/transcript annotations\n",
    "\n",
    "### Reference genome\n",
    "\n",
    "The Mus musculus GRCm38 reference genome in FASTA format was downloaded from [Ensembl](http://nov2020.archive.ensembl.org/Mus_musculus/Info/Index#) via the command ```wget ftp://ftp.ensembl.org:21/pub/release-102/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.primary_assembly.fa.gz```. This is the reference genome that was stated to be used in the paper.\n",
    "\n",
    "Location on disk: ```~/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.dna.primary_assembly.fa```\n",
    "\n",
    "The Mus musculus GRCm39 reference genome in FASTA format was also downloaded ```wget https://ftp.ensembl.org/pub/release-109/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.primary_assembly.fa.gz```, but it was not used in this analysis.\n",
    "\n",
    "Location: ```~/bioinformatics/reference_genomes/GRCm39/Mus_musculus.GRCm39.dna.primary_assembly.fa```\n",
    "\n",
    "### Gene annotations in .gtf format\n",
    "\n",
    "Gene annotations in .gtf format were downloaded from Ensembl as well via the command ```wget ftp://ftp.ensembl.org:21/pub/release-102/gtf/mus_musculus/Mus_musculus.GRCm38.102.gtf.gz```. There are other versions of the .gtf files so I will see if those are needed later on.\n",
    "\n",
    "Location: ```~/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf```\n",
    "\n",
    "I also downloaded the gene annotations in .gtf from Ensembl for GRCm39 via the command ```ftp://ftp.ensembl.org/pub/release-109/gtf/mus_musculus/Mus_musculus.GRCm39.109.gtf.gz```\n",
    "\n",
    "You can preview the file with ```head -n 50 ~/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf | column -t | less -p exon -S```\n",
    "\n",
    "```cat /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf | grep -w gene | wc -l```\n",
    "\n",
    "You can view the structure of one transcript with ```grep ENSMUSG00000102693 /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf | less -p \"exon\\s\" -S```\n",
    "\n",
    "**Important**: chromosome names in the .gtf file must be the same as those in the reference genome .fa file. If StringTie returns expression values that are all 0, this is the likley cause of that error. Download the annotation and reference genome files from the same source.\n",
    "\n",
    "**Make sure your reference genome and annotations are for the same build of the genome (i.e. both for GRCh38).**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "582e596f-2d4f-49c5-9841-ad0958963545",
   "metadata": {},
   "source": [
    "***\n",
    "# Creating index with HISAT2\n",
    "\n",
    "**WARNING** This step will take **A LOT** of RAM, well over 100 gb so make sure the system has enough.\n",
    "\n",
    "It is much easier and faster, if possible, to download a pre-built index provided by the HISAT2 devs at [their website](https://daehwankimlab.github.io/hisat2/download/).\n",
    "\n",
    "I downloaded the GRCm38 index, files were renamed with ```rename genome Mus_musculus.GRCm38.dna.primary_assembly *.ht2```.\n",
    "\n",
    "This was done directly using python files from the [HISAT2 github repo](https://github.com/DaehwanKimLab/hisat2), not sure how to do it using the installed program listed above or if it's possible. Just clone the repo and use the files necessary.\n",
    "\n",
    "Switch to hisat cloned repo ```cd /home/user/bioinformatics/hisat2```\n",
    "\n",
    "Run the following commands.\n",
    "\n",
    "```./hisat2_extract_splice_sites.py ~/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf > ~/bioinformatics/BP_PRJNA496042/splicesites.tsv```\n",
    "\n",
    "```./hisat2_extract_exons.py ~/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf > ~/bioinformatics/BP_PRJNA496042/exons.tsv```\n",
    "\n",
    "```hisat2-build -p 24 --ss ~/bioinformatics/BP_PRJNA496042/splicesites.tsv --exon ~/bioinformatics/BP_PRJNA496042/exons.tsv ~/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.dna.primary_assembly.fa ~/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.dna.primary_assembly```\n",
    "\n",
    "The output should be a collection of binary files with names like Mus_musculus.GRCm38.dna.primary_assembly.x.ht2 or .rf, the x being a number."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11d2c932-2621-462d-892c-c71cc7372b0f",
   "metadata": {},
   "source": [
    "***\n",
    "# Look at some of the actual RNA-seq data\n",
    "\n",
    "Look at the head of the files.\n",
    "\n",
    "```cd ~/bioinformatics/PyMT_Paper_2023```\n",
    "\n",
    "```zcat Pistilli_P11B_FC1_S1_L001_R1_001.fastq.gz | head -n 100```\n",
    "\n",
    "A cursory look shows us that the average read length appears to be around 50 bp. Since these are paired-end reads, that makes sense. For the quality lines, most seem to be ```C```, which corresponds to a Phred quality score of 33, or a 5% chance that the base was incorrectly called.\n",
    "\n",
    "Determine how many reads are in a file.\n",
    "\n",
    "```zcat Pistilli_P11B_FC1_S1_L001_R1_001.fastq.gz | grep -P \"^\\@VH00\" | wc -l```\n",
    "\n",
    "Note, you may have to change the grep command to match the instrument ID.\n",
    "\n",
    "There are 22,898,344 reads in this file.\n",
    "\n",
    "There is a bash file in the Github repo under the PyMT_Paper_2023 folder titled ```total_PE_read_counts.sh``` that will count all the reads for all FASTQ files and output them to a txt file. Simply define the directory for the FASTQ files, the output directory for the text file, and if there is a different naming scheme besides ```R1``` and ```R2``` for the paired reads, change those value sin the script. You may also need to change the search value for grep from ```VH00``` to what is in your FASTQ files.\n",
    "\n",
    "### After looking at the total read counts there are some potential issues\n",
    "\n",
    "Three files returned an 'unexpected end of file' error from gzip. They had only around 10% at most the number of reads of the rest of the files (1-5 million vs 20+ million) Those being:\n",
    "\n",
    "```Pistilli_P13C_FC4_S4_L001_R2_001.fastq.gz```\n",
    "\n",
    "```Pistilli_P188-2_MC5_S15_L001_R2_001.fastq.gz```\n",
    "\n",
    "```Pistilli_P188-4_MT4_S19_L001_R2_001.fastq.gz```\n",
    "\n",
    "I will need to figure out what to do about these. First I will check the source files to make sure they are correctly downloaded. \n",
    "\n",
    "After re-downloading the files, it just seems they were not correctly copied. The full size files have been added to the directory. The results from the bash script are located at ```/home/user/bioinformatics/PyMT_Paper_2023/total_read_counts_2023-02-21_19-15-35.txt```.\n",
    "\n",
    "This script confirms all read pairs are equal in length."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c5becb9-95ab-452b-8c8f-0ee038c30435",
   "metadata": {},
   "source": [
    "***\n",
    "# Check strandedness\n",
    "\n",
    "This step is not required but due to differences in usage of RNA-seq library construction kits, it is **highly** recommended to empirically verify strandedness using the below method.\n",
    "\n",
    "The check_strandedness should already be installed at the beginning. It can also be installed as a [Docker image](https://hub.docker.com/r/mgibio/checkstrandedness).\n",
    "\n",
    "I suppose putting these in the GRCm38 folder should work for now.\n",
    "\n",
    "```/home/user/bioinformatics/reference_genomes/GRCm38```\n",
    "\n",
    "Convert gtf to genepred\n",
    "\n",
    "```cd /home/user/bioinformatics/gtfToGenePred```\n",
    "\n",
    "```./gtfToGenePred /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.genePred```\n",
    "\n",
    "Convert genPred to bed12\n",
    "\n",
    "```cd /home/user/bioinformatics/genePredToBed```\n",
    "\n",
    "```./genePredToBed /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.genePred /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.bed12```\n",
    "\n",
    "Use bedtools to create fasta index from GTF\n",
    "\n",
    "```bedtools getfasta -fi /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.dna.primary_assembly.fa -bed /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.bed12 -s -split -name -fo /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.dna.primary_assembly.transcripts.fa```\n",
    "\n",
    "Clean up the header lines of the transcript file\n",
    "\n",
    "```cat /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.dna.primary_assembly.transcripts.fa | perl -ne 'if($_ =~/^\\>\\S+\\:\\:(ERCC\\-\\d+)\\:.*/){print \">$1\\n\"}elsif ($_ =~/^\\>(\\S+)\\:\\:.*/){print \">$1\\n\"}else{print $_}' > /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.dna.primary_assembly.transcripts.clean.fa```\n",
    "\n",
    "View the clean file\n",
    "\n",
    "```less /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.dna.primary_assembly.transcripts.fa```\n",
    "\n",
    "Add in transcript_id field to .gtf file\n",
    "\n",
    "```awk '{ if ($0 ~ \"transcript_id\") print $0; else print $0\" transcript_id \\\"\\\";\"; }' /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf > /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102_tidy.gtf```\n",
    "\n",
    "check strandedness\n",
    "\n",
    "**Everything up to here should be done already**.\n",
    "\n",
    "```conda activate bioinformatics```\n",
    "\n",
    "This command saves some files to the directory it is run from, best to change to where the fastq files are.\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Paper_2023```\n",
    "\n",
    "```check_strandedness --gtf /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102_tidy.gtf --transcripts /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.dna.primary_assembly.transcripts.clean.fa --reads_1 /home/user/bioinformatics/PyMT_Paper_2023/Pistilli_P11B_FC1_S1_L001_R1_001.fastq.gz --reads_2 /home/user/bioinformatics/PyMT_Paper_2023/Pistilli_P11B_FC1_S1_L001_R2_001.fastq.gz```\n",
    "\n",
    "OUTPUT:\n",
    "\n",
    "Total 200000 usable reads were sampled\n",
    "\n",
    "This is PairEnd Data\n",
    "\n",
    "Fraction of reads failed to determine: 0.1266\n",
    "\n",
    "Fraction of reads explained by \"1++,1--,2+-,2-+\": 0.0249 (2.9% of explainable reads)\n",
    "\n",
    "Fraction of reads explained by \"1+-,1-+,2++,2--\": 0.8485 (97.1% of explainable reads)\n",
    "\n",
    "Over 90% of reads explained by \"1+-,1-+,2++,2--\"\n",
    "\n",
    "Data is likely RF/fr-firststrand\n",
    "\n",
    "More info on strandedness can be found [here](https://rnabio.org/module-09-appendix/0009/12/01/StrandSettings/)\n",
    "\n",
    "Based on the strandedness, this sequencing was likely done using the dUTP method."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "305030d2-d505-454c-a5bf-ae6608fe5d99",
   "metadata": {},
   "source": [
    "***\n",
    "# Pre-alignment QC\n",
    "\n",
    "To get a sense of the raw reads quality in your fastq files, run the following programs.\n",
    "\n",
    "### fastqc [docs](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/)\n",
    "\n",
    "```cd /home/user/bioinformatics/BP_PRJNA496042```\n",
    "\n",
    "```fastqc -t 24 *.fastq.gz``` in the same directory as your reads.\n",
    "\n",
    "It will output an HTML file for each read in the same directory that contains a full report of the quality measures.\n",
    "\n",
    "In the \"over-represented sequences\" section, try [BLASTing](https://blast.ncbi.nlm.nih.gov/Blast.cgi) the sequences to try determining a source organism.\n",
    "\n",
    "### multiqc\n",
    "\n",
    "Run multiqc to combine all the fastqc reports.\n",
    "\n",
    "```cd /home/user/bioinformatics/BP_PRJNA496042```\n",
    "\n",
    "Run ```multiqc .``` in the same directory as the reports.\n",
    "\n",
    "Renamed file to ```pre-alignment_multiqc_report.html```\n",
    "\n",
    "Make a new directory for the fastqc .html outputs and move them\n",
    "\n",
    "```mkdir fastqc && mv *_fastqc* fastqc```\n",
    "\n",
    "## Bash script\n",
    "\n",
    "I made a bash script to handle this step, ```pre-alignment_QC.sh```. All that should need to be changed is the fastq_dir variable. Which should point to where the FASTQ files are. The script does not yet account for any sub-folders.\n",
    "\n",
    "## View reports\n",
    "\n",
    "Looking at the MultiQC report, none of the warnings are concerning or unusual. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "334baa6d-01d5-4841-b232-53c9ab626c0d",
   "metadata": {},
   "source": [
    "***\n",
    "# Remove adapters and trim\n",
    "\n",
    "Make a new folder for the trimmed reads\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Paper_2023```\n",
    "\n",
    "```mkdir trimmed && cd trimmed```\n",
    "\n",
    "Download the Illumina adapter sequence files, or the relevant files for your sequencing machine.\n",
    "\n",
    "```wget http://genomedata.org/rnaseq-tutorial/illumina_multiplex.fa```\n",
    "\n",
    "Downloaded to ```/home/user/bioinformatics/reference_genomes/GRCm38/illumina_multiplex.fa```\n",
    "\n",
    "Use flexbar to remove adapter sequences and trim first 13 bases of each read. Adjust the number of threads to something manageable for your system. It is set to 24 threads but may take much longer with a slower CPU.\n",
    "\n",
    "Possible options for flexbar and their meaning:\n",
    "\n",
    "   - ```–adapter-min-overlap 7``` requires a minimum of 7 bases to match the adapter\n",
    "   - ```–adapter-trim-end RIGHT``` uses a trimming strategy to remove the adapter from the 3 prime or RIGHT end of the read\n",
    "   - ```–max-uncalled 300``` allows as many as 300 uncalled or N bases (MiSeq read lengths can be 300bp)\n",
    "   - ```–min-read-length``` the minimum read length allowed after trimming is 25bp.\n",
    "   - ```–threads 24``` use 24 threads\n",
    "   - ```–zip-output GZ``` the input FASTQ files are gzipped so we will output gzipped FASTQ to save space\n",
    "   - ```–adapters``` define the path to the adapter FASTA file to trim\n",
    "   - ```–reads``` define the path to the read 1 FASTQ file of reads\n",
    "   - ```–reads2``` define the path to the read 2 FASTQ file of reads\n",
    "   - ```–target``` a base path for the output files. The value will _1.fastq.gz and _2.fastq.gz for read 1 and read 2 respectively\n",
    "   - ```–pre-trim-left``` trim a fixed number of bases at left read end. For example, to trim 5 bases at the left side of reads: –pre-trim-left 5\n",
    "   - ```–pre-trim-right``` trim a fixed number of bases at right read end. For example, to trim 5 bases at the right side of reads: –pre-trim-right 5\n",
    "   - ```–pre-trim-phred``` trim based on phred quality value to deal with higher error rates towards the end of reads. For example, to trim the 3 prime end until quality offset value 30 or higher is reached, specify: –pre-trim-phred 30\n",
    "\n",
    "The commands to run for flexbar for this dataset are as follows:\n",
    "\n",
    "```cd /home/user/bioinformatics/flexbar-3.5.0-linux```\n",
    "\n",
    "You will need to export the library path environment variable for this session, or add to shell startup scrits.\n",
    "\n",
    "```export LD_LIBRARY_PATH=/home/user/bioinformatics/flexbar-3.5.0-linux:$LD_LIBRARY_PATH```\n",
    "\n",
    "```/home/user/bioinformatics/flexbar-3.5.0-linux/flexbar --adapter-min-overlap 7 --adapter-trim-end RIGHT --adapters /home/user/bioinformatics/reference_genomes/GRCm38/illumina_multiplex.fa --pre-trim-left 13 --max-uncalled 300 --min-read-length 25 --threads 24 --zip-output GZ --reads /home/user/bioinformatics/PyMT_Paper_2023/Pistilli_P11B_FC1_S1_L001_001.read1.fastq.gz --reads2 /home/user/bioinformatics/PyMT_Paper_2023/Pistilli_P11B_FC1_S1_L001_001.read2.fastq.gz --target /home/user/bioinformatics/PyMT_Paper_2023/trimmed/Pistilli_P11B_FC1_S1_L001_001```\n",
    "\n",
    "The rest of the commands were generated using the script ```remove_adapters_and_trim.sh```, but they all follow this same format.\n",
    "\n",
    "Compare the impact of trimming with the original QC reports by running fastqc and multiqc on the trimmed reads\n",
    "\n",
    "In the directory of the trimmed reads ```cd /home/user/bioinformatics/PyMT_Paper_2023/trimmed```, run:\n",
    "\n",
    "```fastqc *.fastq.gz```\n",
    "\n",
    "```multiqc .```\n",
    "\n",
    "Same as before, make a new directory for the fastqc .html outputs and move them\n",
    "\n",
    "```mkdir fastqc && mv *_fastqc* fastqc```\n",
    "\n",
    "A script for the QC was also written, ```post-trim_QC.sh```\n",
    "\n",
    "Looking at the MultiQC report after trimming, the overall quality looks better. Less failures and warnings. Mostly just sequence duplication now, which is expected.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74b597a8-7053-43b2-9bc2-885281ec1894",
   "metadata": {},
   "source": [
    "***\n",
    "# Alignment with HISAT2\n",
    "\n",
    "For more details on HISAT2, refer to the [manual on Github](http://daehwankimlab.github.io/hisat2/manual/)\n",
    "\n",
    "The outputs after running HISAT2 will be a SAM/BAM file for each data set.\n",
    "\n",
    "Options for HISAT2 include:\n",
    "\n",
    "\n",
    "   - ```-p 24``` tells HISAT2 to use 24 threads for bowtie alignments.\n",
    "   - ```–rna-strandness``` specifies strandness of RNAseq library. Check_strandedness said it is likely unstranded, which is the default. So we remove this flag in this run.\n",
    "   - ```–rg-id $ID``` specifies a read group ID that is a unique identifier.\n",
    "   - ```–rg SM:$SAMPLE_NAME``` specifies a read group sample name. This together with rg-id will allow you to determine which reads came from which sample in the merged bam later on.\n",
    "   - ```–rg LB:$LIBRARY_NAME``` specifies a read group library name. This together with rg-id will allow you to determine which reads came from which library in the merged bam later on.\n",
    "   - ```–rg PL:ILLUMINA``` specifies a read group sequencing platform.\n",
    "   - ```–rg PU:$PLATFORM_UNIT``` specifies a read group sequencing platform unit. Typically this consists of FLOWCELL-BARCODE.LANE\n",
    "   - ```–dta``` Reports alignments tailored for transcript assemblers.\n",
    "   - ```-x /path/to/hisat2/index``` The HISAT2 index filename prefix (minus the trailing .X.ht2) built earlier including splice sites and exons.\n",
    "   - ```-1 /path/to/read1.fastq.gz``` The read 1 FASTQ file, optionally gzip(.gz) or bzip2(.bz2) compressed.\n",
    "   - ```-2 /path/to/read2.fastq.gz``` The read 2 FASTQ file, optionally gzip(.gz) or bzip2(.bz2) compressed.\n",
    "   - ```-U /path/to/unstranded.fastq.gz``` If data is unstranded, provide link to FASTQ or .gz version here.\n",
    "   - ```-S /path/to/output.sam``` The output SAM format text file of alignments.\n",
    "   \n",
    "Create an alignment folder for the outputs.\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Paper_2023 && mkdir -p alignments && cd alignments```\n",
    "\n",
    "Run a command for each of the replicates. I will try running based off the trimmed fastq files and see how it goes.\n",
    "\n",
    "I wrote a script to handle this, ```HISAT2_alignment.sh```\n",
    "\n",
    "```hisat2 -p 24 --rg-id Pistilli_P11B_FC1_S1_L001_001 --rg SM:Pistilli_P11B_FC1_S1_L001_001 --rg LB:Pistilli_P11B_FC1_S1_L001_001_Lib --rg PL:ILLUMINA -x /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.dna.primary_assembly --dta -1 /home/user/bioinformatics/PyMT_Paper_2023/trimmed/Pistilli_P11B_FC1_S1_L001_001_1.fastq.gz -2 /home/user/bioinformatics/PyMT_Paper_2023/trimmed/Pistilli_P11B_FC1_S1_L001_001_2.fastq.gz -S /home/user/bioinformatics/PyMT_Paper_2023/trimmed/Pistilli_P11B_FC1_S1_L001_001.sam```\n",
    "\n",
    "\n",
    "\n",
    "### Convert SAM to BAM files\n",
    "\n",
    "I also wrote a script for this, ```Convert_SAM_to_BAM.sh```\n",
    "\n",
    "```-@ int``` is the number of threads to use.\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Paper_2023/alignments```\n",
    "\n",
    "```samtools sort -@ 24 -o Pistilli_P11B_FC1_S1_L001_001.bam Pistilli_P11B_FC1_S1_L001_001.sam```\n",
    "\n",
    "And so on so forth...\n",
    "\n",
    "### Merge all BAM files\n",
    "Other tools such as \"samtools merge\" and \"bamtools merge\" can also achieve this. Create a BAM file for each group (FC, FT, MC, MT).\n",
    "\n",
    "```java -Xmx2g -jar /home/user/bioinformatics/picard.jar MergeSamFiles -OUTPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/FC.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P11B_FC1_S1_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P11D_FC2_S2_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P13B_FC3_S3_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P13C_FC4_S4_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P13E_FC5_S5_L001_001.bam```\n",
    "\n",
    "```java -Xmx2g -jar /home/user/bioinformatics/picard.jar MergeSamFiles -OUTPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/FT.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P12BL_FT1_S6_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P12CRR_FT2_S7_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P13A_FT3_S8_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P14E_FT4_S9_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P13F_FT5_S10_L001_001.bam```\n",
    "\n",
    "```java -Xmx2g -jar /home/user/bioinformatics/picard.jar MergeSamFiles -OUTPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/MC.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P186-5_MC1_S11_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P186-7_MC2_S12_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P189-1_MC3_S13_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P187-1_MC4_S14_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P188-2_MC5_S15_L001_001.bam```\n",
    "\n",
    "```java -Xmx2g -jar /home/user/bioinformatics/picard.jar MergeSamFiles -OUTPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/MT.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P187-3_MT1_S16_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P186-6_MT2_S17_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P188-1_MT3_S18_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P188-4_MT4_S19_L001_001.bam -INPUT /home/user/bioinformatics/PyMT_Paper_2023/alignments/Pistilli_P188-5_MT5_S20_L001_001.bam```\n",
    "\n",
    "You can verify that all bam files were created successfully, there should be 24.\n",
    "\n",
    "In the same directory as the aligned .bam files, ```ls -l *.bam | wc -l```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "868c581b-09b0-4ea6-8ca4-24cf0be5ed81",
   "metadata": {},
   "source": [
    "***\n",
    "# Using the Integrative Genomics Viewer (IGV)\n",
    "\n",
    "```cd /home/user/bioinformatics/IGV_Linux_2.15.4```\n",
    "\n",
    "IGV can be run from the download directory using ```./igv.sh```\n",
    "\n",
    "To re-visit the shortish tutorial on navigating IGV, go [here](https://rnabio.org/module-02-alignment/0002/04/01/IGV/). You can follow the tutorial there using the provided files.\n",
    "\n",
    "For our dataset, we need to index the BAM files before loading them into IGV. This will be done with samtools. From the directory where the BAM files are located, run:\n",
    "\n",
    "You can just run the following command to index all .bam files in the directory. This way is easier than running multiple commands.\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Paper_2023/alignments```\n",
    "\n",
    "```find *.bam -exec echo samtools index {} \\; | sh```\n",
    "\n",
    "But in this case, since I don't want to index all the individual sample files, I will just do the joined files.\n",
    "\n",
    "```find FC.bam FT.bam MC.bam MT.bam -exec echo samtools index {} \\; | sh```\n",
    "\n",
    "Load both the .bam files in IGV and look around. The .bam.bai index needs to be in the same directory as the .bam files."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b31f311f-bd5e-4410-83b6-1ee1ee71d00d",
   "metadata": {},
   "source": [
    "***\n",
    "# Alignment QC\n",
    "\n",
    "Use samtools to view the format of the alignment bam files.\n",
    "\n",
    "```samtools view /home/user/bioinformatics/PyMT_Paper_2023/alignments/FC.bam | head | column -t | less -S```\n",
    "\n",
    "```samtools view /home/user/bioinformatics/PyMT_Paper_2023/alignments/FT.bam | head | column -t | less -S```\n",
    "\n",
    "```samtools view /home/user/bioinformatics/PyMT_Paper_2023/alignments/MC.bam | head | column -t | less -S```\n",
    "\n",
    "```samtools view /home/user/bioinformatics/PyMT_Paper_2023/alignments/MT.bam | head | column -t | less -S```\n",
    "\n",
    "You can add additional filtering flags to the samtools command. [This useful guide explains all of them](https://broadinstitute.github.io/picard/explain-flags.html). ```-f``` are required flags and ```-F``` are filtering flags.\n",
    "\n",
    "Look for alignments that are only for 'PCR or optical duplicates'.\n",
    "\n",
    "```samtools view -f 1024 /home/user/bioinformatics/PyMT_Paper_2023/alignments/FC.bam | head```\n",
    "\n",
    "Output: nothing\n",
    "\n",
    "The command ```samtools flagstat``` can provide a summary of the alignments.\n",
    "\n",
    "```mkdir -p /home/user/bioinformatics/PyMT_Paper_2023/alignments/flagstat```\n",
    "\n",
    "```cd /home/user/bioinformatics/PyMT_Paper_2023/alignments```\n",
    "\n",
    "```samtools flagstat /home/user/bioinformatics/BP_PRJNA496042/alignments/Control_1.bam > /home/user/bioinformatics/BP_PRJNA496042/alignments/flagstat/Control_1.bam.flagstat```\n",
    "\n",
    "```samtools flagstat /home/user/bioinformatics/BP_PRJNA496042/alignments/Control_2.bam > /home/user/bioinformatics/BP_PRJNA496042/alignments/flagstat/Control_2.bam.flagstat```\n",
    "\n",
    "```samtools flagstat /home/user/bioinformatics/BP_PRJNA496042/alignments/Control_3.bam > /home/user/bioinformatics/BP_PRJNA496042/alignments/flagstat/Control_3.bam.flagstat```\n",
    "\n",
    "```samtools flagstat /home/user/bioinformatics/BP_PRJNA496042/alignments/Control_4.bam > /home/user/bioinformatics/BP_PRJNA496042/alignments/flagstat/Control_4.bam.flagstat```\n",
    "\n",
    "```samtools flagstat /home/user/bioinformatics/BP_PRJNA496042/alignments/PDX_1.bam > /home/user/bioinformatics/BP_PRJNA496042/alignments/flagstat/PDX_1.bam.flagstat```\n",
    "\n",
    "```samtools flagstat /home/user/bioinformatics/BP_PRJNA496042/alignments/PDX_2.bam > /home/user/bioinformatics/BP_PRJNA496042/alignments/flagstat/PDX_2.bam.flagstat```\n",
    "\n",
    "```samtools flagstat /home/user/bioinformatics/BP_PRJNA496042/alignments/PDX_3.bam > /home/user/bioinformatics/BP_PRJNA496042/alignments/flagstat/PDX_3.bam.flagstat```\n",
    "\n",
    "```samtools flagstat /home/user/bioinformatics/BP_PRJNA496042/alignments/PDX_4.bam > /home/user/bioinformatics/BP_PRJNA496042/alignments/flagstat/PDX_4.bam.flagstat```\n",
    "\n",
    "You can also run an expression to do it in one command.\n",
    "\n",
    "```find *_*.bam -exec echo samtools flagstat {} \\> flagstat/{}.flagstat \\; | sh```\n",
    "\n",
    "### FastQc on bam file\n",
    "\n",
    "From your alignment directory (where the bam files are).\n",
    "\n",
    "```mkdir /home/user/bioinformatics/BP_PRJNA496042/alignments/fastqc```\n",
    "\n",
    "```cd /home/user/bioinformatics/BP_PRJNA496042/alignments```\n",
    "\n",
    "```fastqc Control_*.bam PDX_*.bam -t 24```\n",
    "\n",
    "Clean up the current directory and move the reports into the fastqc folder.\n",
    "\n",
    "```mv *fastqc.html fastqc/ && mv *fastqc.zip fastqc/```\n",
    "\n",
    "### Picard\n",
    "\n",
    "Using Picard we can generate metrics and figures that express the quality of our RNA-seq data.\n",
    "\n",
    "We first need to create input files for Picard based on the reference genome so it can properly run the function CollectRnaSeqMetrics.\n",
    "\n",
    "```java -jar /home/user/bioinformatics/picard.jar CreateSequenceDictionary -R /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.dna.primary_assembly.fa -O /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.dna.primary_assembly.dict```\n",
    "\n",
    "Create a bed file of the location of ribosomal sequences in our reference (first extract from the gtf then convert to bed). Note that here we pull all the \"rrna\" transcripts from the GTF.\n",
    "\n",
    "```grep --color=none -i -P \"rrna\" /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf > /home/user/bioinformatics/reference_genomes/GRCm38/ref_ribosome.gtf```\n",
    "\n",
    "Note: ```gff2bed``` is part of BEDOPS. Run the script directly from the bedops/bin folder. You may need to add the /bin folder to your path variable ```export PATH=\"$PATH:/home/user/bioinformatics/bedops_linux_x86_64-v2.4.40/bin\"```, otherwise convert2bed may not be found.\n",
    "\n",
    "```cd /home/user/bioinformatics/bedops_linux_x86_64-v2.4.40/bin```\n",
    "\n",
    "```gff2bed < /home/user/bioinformatics/reference_genomes/GRCm38/ref_ribosome.gtf > /home/user/bioinformatics/reference_genomes/GRCm38/ref_ribosome.bed```\n",
    "\n",
    "Create interval list file to identify the location of the ribosomal sequences in the reference\n",
    "\n",
    "```java -jar /home/user/bioinformatics/picard.jar BedToIntervalList -I /home/user/bioinformatics/reference_genomes/GRCm38/ref_ribosome.bed -O /home/user/bioinformatics/reference_genomes/GRCm38/ref_ribosome.interval_list -SD /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.dna.primary_assembly.dict```\n",
    "\n",
    "Create genePred file for the reference transcriptome\n",
    "\n",
    "```cd /home/user/bioinformatics/gtfToGenePred```\n",
    "\n",
    "```./gtfToGenePred -genePredExt /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.ref_flat.txt```\n",
    "\n",
    "Reformat the genePred file\n",
    "\n",
    "```cat /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.ref_flat.txt | awk '{print $12\"\\t\"$0}' | cut -d$'\\t' -f1-11 > /home/user/bioinformatics/reference_genomes/GRCm38/tmp.txt```\n",
    "\n",
    "```mv /home/user/bioinformatics/reference_genomes/GRCm38/tmp.txt /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.ref_flat.txt```\n",
    "\n",
    "Go to the directory with your alignments.\n",
    "\n",
    "```cd /home/user/bioinformatics/BP_PRJNA496042/alignments && mkdir picard```\n",
    "\n",
    "Since this data is unstranded, I removed the ```STRAND=SECOND_READ_TRANSCRIPTION_STRAND``` flag since half the reads were not mapped correctly in the quality report. Instead it was set to none. \n",
    "\n",
    "```find *_*.bam -exec echo java -jar /home/user/bioinformatics/picard.jar CollectRnaSeqMetrics I={} O=picard/{}.RNA_Metrics REF_FLAT=/home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.ref_flat.txt STRAND=NONE RIBOSOMAL_INTERVALS=/home/user/bioinformatics/reference_genomes/GRCm38/ref_ribosome.interval_list \\; | sh```\n",
    "\n",
    "### RSeQC\n",
    "\n",
    "Another tool for generating RNA-seq QC reports.\n",
    "\n",
    "You will need\n",
    "- aligned bam files\n",
    "- index file for each bam file\n",
    "- transcript bed file (in bed12 format)\n",
    "\n",
    "Start by converting the reference gtf to genePred. This step has probably already been done.\n",
    "\n",
    "```gtfToGenePred /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.genePred```\n",
    "\n",
    "Then convert the genePred to bed12. This step also probably done already.\n",
    "\n",
    "```genePredToBed /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.genePred /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.bed12```\n",
    "\n",
    "Go back to the alignment directory. Make sure the bioinformatics env is active.\n",
    "\n",
    "```cd /home/user/bioinformatics/BP_PRJNA496042/alignments && mkdir rseqc```\n",
    "\n",
    "```geneBody_coverage.py -i Control_1.bam, Control_2.bam, Control_3.bam, Control_4.bam -r /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.bed12 -o rseqc/CONTROL```\n",
    "\n",
    "```geneBody_coverage.py -i PDX_1.bam, PDX_2.bam, PDX_3.bam, PDX_4.bam -r /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.bed12 -o rseqc/PDX```\n",
    "\n",
    "The following commands should also be run from the alignment directory ```cd /home/user/bioinformatics/BP_PRJNA496042/alignments```\n",
    "\n",
    "```find *_*.bam -exec echo inner_distance.py -i {} -r /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.bed12 -o rseqc/{} \\; | sh```\n",
    "\n",
    "```find *_*.bam -exec echo junction_annotation.py -i {} -r /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.bed12 -o rseqc/{} \\; | sh```\n",
    "\n",
    "```find *_*.bam -exec echo junction_saturation.py -i {} -r /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.bed12 -o rseqc/{} \\; | sh```\n",
    "\n",
    "```find *_*.bam -exec echo read_distribution.py -i {} -r /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.bed12 \\> rseqc/{}.read_dist.txt \\; | sh```\n",
    "\n",
    "```find *_*.bam -exec echo RNA_fragment_size.py -i {} -r /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.bed12 \\> rseqc/{}.frag_size.txt \\; | sh```\n",
    "\n",
    "```find *_*.bam -exec echo bam_stat.py -i {} \\> {}.bam_stat.txt \\; | sh```\n",
    "\n",
    "```rm -f log.txt```\n",
    "\n",
    "### MultiQC\n",
    "\n",
    "Now we will compile a multiQC report from all of the QC done above.\n",
    "\n",
    "In the alignment directory ```cd /home/user/bioinformatics/BP_PRJNA496042/alignments```.\n",
    "\n",
    "```multiqc ./```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eab94d3a-47d6-435d-a4d6-19687fe7d5b8",
   "metadata": {},
   "source": [
    "***\n",
    "# Expression Analysis\n",
    "\n",
    "Some notes before we start.\n",
    "\n",
    "First, there are multiple things you can do with your aligned reads from HISAT2. This pipeline is written with the intention of going through the differential expression analysis and visualizing it. But, you can also just use **HTSEQ-count** to get raw read counts instead of FPKM values. \n",
    "\n",
    "FPKM = fragments per kilobase of transcript per million mapped reads. FPKM attempts to normalize for gene size and library depth. RPKM is essentially the same thing, however there are differing opinions on just how similar they are, and which is better. \n",
    "\n",
    "Mathematically, FPKM can be represented as:\n",
    "\n",
    "```FPKM = (C / (N / 1000000)) / (L / 1000)```, where C = number of mappable fragments for a gene, N = total number of mappable fragments in the library, and L = number of base pairs in the gene.\n",
    "\n",
    "Another normalization metric that may be mentioned is TPM (transcript per kilobase million). The only difference between FPKM and TPM is the order of operations. At least, in the calculation of it.\n",
    "\n",
    "### Stringtie\n",
    "\n",
    "If you want to know more about the math behind how Stringtie works, you can [read the paper](https://www.nature.com/articles/nbt.3122).\n",
    "\n",
    "An outline of the Stringtie flags used:\n",
    "\n",
    "\n",
    "- ```–rf``` tells StringTie that our data is stranded and to use the correct strand specific mode (i.e. assume a stranded library fr-firststrand). In this case, our data is unstranded so this fla g is not used.\n",
    "- ```-p 24``` tells StringTie to use 24 threads.\n",
    "- ```-G``` reference annotation to use for guiding the assembly process (GTF/GFF3)\n",
    "- ```-e``` only estimate the abundance of given reference transcripts (requires -G)\n",
    "- ```-B``` enable output of Ballgown table files which will be created in the same directory as the output GTF (requires -G, -o recommended)\n",
    "- ```-o``` output path/file name for the assembled transcripts GTF (default: stdout)\n",
    "- ```-A``` output path/file name for gene abundance estimates\n",
    "\n",
    "I'll make this folder within the alignment folder ```cd /home/user/bioinformatics/BP_PRJNA496042/alignments```.\n",
    "\n",
    "```mkdir -p expression/stringtie/ref_only/ && cd expression/stringtie/ref_only/```\n",
    "\n",
    "```cd /home/user/bioinformatics/stringtie-2.2.1.Linux_x86_64```\n",
    "\n",
    "```./stringtie -p 24 -G /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf -e -B -o /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/Control_1/transcripts.gtf -A /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/Control_1/gene_abundances.tsv /home/user/bioinformatics/BP_PRJNA496042/alignments/Control_1.bam```\n",
    "\n",
    "```./stringtie -p 24 -G /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf -e -B -o /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/Control_2/transcripts.gtf -A /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/Control_2/gene_abundances.tsv /home/user/bioinformatics/BP_PRJNA496042/alignments/Control_2.bam```\n",
    "\n",
    "```./stringtie -p 24 -G /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf -e -B -o /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/Control_3/transcripts.gtf -A /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/Control_3/gene_abundances.tsv /home/user/bioinformatics/BP_PRJNA496042/alignments/Control_3.bam```\n",
    "\n",
    "```./stringtie -p 24 -G /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf -e -B -o /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/Control_4/transcripts.gtf -A /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/Control_4/gene_abundances.tsv /home/user/bioinformatics/BP_PRJNA496042/alignments/Control_4.bam```\n",
    "\n",
    "```./stringtie -p 24 -G /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf -e -B -o /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/PDX_1/transcripts.gtf -A /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/PDX_1/gene_abundances.tsv /home/user/bioinformatics/BP_PRJNA496042/alignments/PDX_1.bam```\n",
    "\n",
    "```./stringtie -p 24 -G /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf -e -B -o /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/PDX_2/transcripts.gtf -A /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/PDX_2/gene_abundances.tsv /home/user/bioinformatics/BP_PRJNA496042/alignments/PDX_2.bam```\n",
    "\n",
    "```./stringtie -p 24 -G /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf -e -B -o /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/PDX_3/transcripts.gtf -A /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/PDX_3/gene_abundances.tsv /home/user/bioinformatics/BP_PRJNA496042/alignments/PDX_3.bam```\n",
    "\n",
    "```./stringtie -p 24 -G /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf -e -B -o /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/PDX_4/transcripts.gtf -A /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/PDX_4/gene_abundances.tsv /home/user/bioinformatics/BP_PRJNA496042/alignments/PDX_4.bam```\n",
    "\n",
    "\n",
    "Have a look at what the raw output from StringTie looks like.\n",
    "\n",
    "```less -S /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/Control_1/transcripts.gtf```\n",
    "\n",
    "Improve formatting of the *less* output.\n",
    "\n",
    "```grep -v \"^#\" /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/Control_1/transcripts.gtf | grep -w \"transcript\" | column -t | less -S```\n",
    "\n",
    "You can also limit the view to just the transcript records and the correspoding FPKM values.\n",
    "\n",
    "```awk '{if ($3==\"transcript\") print}' /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/Control_1/transcripts.gtf | cut -f 1,4,9 | column -t | less -S```\n",
    "\n",
    "You can also view the transcript level expression values for each set of reads in the two files ```t_data.ctab``` and ```gene_abundances.tsv```. They should be located in the folders ```~/expression/stringtie/ref_only/``` and then in the Control or PDX folders.\n",
    "\n",
    "Download the script for creating an expression matrix based off the Stringtie outputs.\n",
    "\n",
    "```cd /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only```\n",
    "\n",
    "```wget https://raw.githubusercontent.com/griffithlab/rnabio.org/master/assets/scripts/stringtie_expression_matrix.pl```\n",
    "\n",
    "```chmod +x stringtie_expression_matrix.pl```\n",
    "\n",
    "```./stringtie_expression_matrix.pl --expression_metric=TPM --result_dirs='Control_1,Control_2,Control_3,Control_4,PDX_1,PDX_2,PDX_3,PDX_4' --transcript_matrix_file=transcript_tpm_all_samples.tsv --gene_matrix_file=gene_tpm_all_samples.tsv```\n",
    "\n",
    "```./stringtie_expression_matrix.pl --expression_metric=FPKM --result_dirs='Control_1,Control_2,Control_3,Control_4,PDX_1,PDX_2,PDX_3,PDX_4' --transcript_matrix_file=transcript_fpkm_all_samples.tsv --gene_matrix_file=gene_fpkm_all_samples.tsv```\n",
    "\n",
    "```./stringtie_expression_matrix.pl --expression_metric=Coverage --result_dirs='Control_1,Control_2,Control_3,Control_4,PDX_1,PDX_2,PDX_3,PDX_4' --transcript_matrix_file=transcript_coverage_all_samples.tsv --gene_matrix_file=gene_coverage_all_samples.tsv```\n",
    "\n",
    "View the TPM output files.\n",
    "\n",
    "```column -t transcript_tpm_all_samples.tsv | less -S```\n",
    "\n",
    "```column -t gene_tpm_all_samples.tsv | less -S```\n",
    "\n",
    "### Generate raw counts instead of FPKM/TPM values\n",
    "\n",
    "If you only want to generate the raw counts this can be done using HTSeq-count [docs](https://htseq.readthedocs.io/).\n",
    "\n",
    "Some notes about the flags used:\n",
    "\n",
    "\n",
    "- ```–format``` specify the input file format one of BAM or SAM. Since we have BAM format files, select ‘bam’ for this option.\n",
    "- ```–order``` provide the expected sort order of the input file. Previously we generated position sorted BAM files so use ‘pos’.\n",
    "- ```–mode``` determines how to deal with reads that overlap more than one feature. We believe the ‘intersection-strict’ mode is best.\n",
    "- ```–stranded``` specifies whether data is stranded or not. The TruSeq strand-specific RNA libraries suggest the ‘reverse’ option for this parameter. This data is unstranded so the flag is set to 'no'.\n",
    "- ```–minaqual``` will skip all reads with alignment quality lower than the given minimum value.\n",
    "- ```–type``` specifies the feature type (3rd column in GFF file) to be used. (default, suitable for RNA-Seq and Ensembl GTF files: exon)\n",
    "- ```–idattr``` The feature ID used to identify the counts in the output table. The default, suitable for RNA-SEq and Ensembl GTF files, is gene_id.\n",
    "\n",
    "Change directories to where the expression folder is ```cd /home/user/bioinformatics/BP_PRJNA496042/alignments```\n",
    "\n",
    "```mkdir -p expression/htseq_counts && cd expression/htseq_counts```\n",
    "\n",
    "Run each htseq-count command to get the raw counts for each read file.\n",
    "\n",
    "```htseq-count --format bam --order pos --mode intersection-strict --stranded no --minaqual 1 --type exon --idattr gene_id /home/user/bioinformatics/BP_PRJNA496042/alignments/Control_1.bam /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf > /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/htseq_counts/Control_1_gene.tsv```\n",
    "\n",
    "```htseq-count --format bam --order pos --mode intersection-strict --stranded no --minaqual 1 --type exon --idattr gene_id /home/user/bioinformatics/BP_PRJNA496042/alignments/Control_2.bam /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf > /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/htseq_counts/Control_2_gene.tsv```\n",
    "\n",
    "```htseq-count --format bam --order pos --mode intersection-strict --stranded no --minaqual 1 --type exon --idattr gene_id /home/user/bioinformatics/BP_PRJNA496042/alignments/Control_3.bam /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf > /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/htseq_counts/Control_3_gene.tsv```\n",
    "\n",
    "```htseq-count --format bam --order pos --mode intersection-strict --stranded no --minaqual 1 --type exon --idattr gene_id /home/user/bioinformatics/BP_PRJNA496042/alignments/Control_4.bam /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf > /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/htseq_counts/Control_4_gene.tsv```\n",
    "\n",
    "```htseq-count --format bam --order pos --mode intersection-strict --stranded no --minaqual 1 --type exon --idattr gene_id /home/user/bioinformatics/BP_PRJNA496042/alignments/PDX_1.bam /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf > /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/htseq_counts/PDX_1_gene.tsv```\n",
    "\n",
    "```htseq-count --format bam --order pos --mode intersection-strict --stranded no --minaqual 1 --type exon --idattr gene_id /home/user/bioinformatics/BP_PRJNA496042/alignments/PDX_2.bam /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf > /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/htseq_counts/PDX_2_gene.tsv```\n",
    "\n",
    "```htseq-count --format bam --order pos --mode intersection-strict --stranded no --minaqual 1 --type exon --idattr gene_id /home/user/bioinformatics/BP_PRJNA496042/alignments/PDX_3.bam /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf > /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/htseq_counts/PDX_3_gene.tsv```\n",
    "\n",
    "```htseq-count --format bam --order pos --mode intersection-strict --stranded no --minaqual 1 --type exon --idattr gene_id /home/user/bioinformatics/BP_PRJNA496042/alignments/PDX_4.bam /home/user/bioinformatics/reference_genomes/GRCm38/Mus_musculus.GRCm38.102.gtf > /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/htseq_counts/PDX_4_gene.tsv```\n",
    "\n",
    "I don't feel like paraphrasing this next section:\n",
    "\n",
    "\"Merge results files into a single matrix for use in edgeR. The following joins the results for each replicate together, adds a header, reformats the result as a tab delimited file, and shows you the first 10 lines of the resulting file :\"\n",
    "\n",
    "```cd /home/user/bioinformatics/BP_PRJNA496042/alignments/expression/htseq_counts```\n",
    "\n",
    "```join Control_1_gene.tsv Control_2_gene.tsv | join - Control_3_gene.tsv | join - Control_4_gene.tsv | join - PDX_1_gene.tsv | join - PDX_2_gene.tsv | join - PDX_3_gene.tsv | join - PDX_4_gene.tsv > gene_read_counts_table_all.tsv```\n",
    "\n",
    "```echo \"GeneID Control_1 Control_2 Control_3 Control_4 PDX_1 PDX_2 PDX_3 PDX_4\" > header.txt```\n",
    "\n",
    "```cat header.txt gene_read_counts_table_all.tsv | grep -v \"__\" | awk -v OFS=\"\\t\" '$1=$1' > gene_read_counts_table_all_final.tsv```\n",
    "\n",
    "If you're curious, ```grep -v \"__\"``` is used to filter out summary lines at the end of htseq-count files that did not align, had poor quality alignment, etc.\n",
    "\n",
    "```rm -f gene_read_counts_table_all.tsv header.txt```\n",
    "\n",
    "```head gene_read_counts_table_all_final.tsv | column -t```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "445456e1-d058-4780-97ea-685a95845bc8",
   "metadata": {},
   "source": [
    "***\n",
    "# Differential Expression (Ballgown)\n",
    "\n",
    "The differential expression is done using the bioconductor package [Ballgown](https://www.bioconductor.org/packages/release/bioc/html/ballgown.html).\n",
    "\n",
    "```cd /home/user/bioinformatics/BP_PRJNA496042/alignments```\n",
    "\n",
    "```mkdir -p de/ballgown/ref_only/ && cd de/ballgown/ref_only/```\n",
    "\n",
    "Start an R session by typing ```R``` in a terminal.\n",
    "\n",
    "The following code is an R script that will perform DE using Ballgown. You can run each command line by line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "68a845c2-34ce-42b3-986e-769ce4aa4715",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘ballgown’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    structure\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:ballgown’:\n",
      "\n",
      "    contains, expr, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: usethis\n",
      "\n",
      "Fri Feb 17 23:41:07 2023\n",
      "\n",
      "Fri Feb 17 23:41:07 2023: Reading linking tables\n",
      "\n",
      "Fri Feb 17 23:41:08 2023: Reading intron data files\n",
      "\n",
      "Fri Feb 17 23:41:10 2023: Merging intron data\n",
      "\n",
      "Fri Feb 17 23:41:11 2023: Reading exon data files\n",
      "\n",
      "Fri Feb 17 23:41:17 2023: Merging exon data\n",
      "\n",
      "Fri Feb 17 23:41:19 2023: Reading transcript data files\n",
      "\n",
      "Fri Feb 17 23:41:21 2023: Merging transcript data\n",
      "\n",
      "Wrapping up the results\n",
      "\n",
      "Fri Feb 17 23:41:22 2023\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "ballgown instance with 142699 transcripts and 8 samples"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load the required libraries\n",
    "library(ballgown)\n",
    "library(genefilter)\n",
    "library(dplyr)\n",
    "library(devtools)\n",
    "\n",
    "# Create phenotype data needed for ballgown analysis\n",
    "ids=c(\"Control_1\",\"Control_2\",\"Control_3\",\"Control_4\",\"PDX_1\",\"PDX_2\",\"PDX_3\",\"PDX_4\")\n",
    "type=c(\"Control\",\"Control\",\"Control\",\"Control\",\"PDX\",\"PDX\",\"PDX\",\"PDX\")\n",
    "results=\"/home/user/bioinformatics/BP_PRJNA496042/alignments/expression/stringtie/ref_only/\" # be sure to include the trailing \"/\"\n",
    "path=paste(results,ids,sep=\"\")\n",
    "pheno_data=data.frame(ids,type,path)\n",
    "\n",
    "# Load ballgown data structure and save it to a variable \"bg\"\n",
    "bg = ballgown(samples=as.vector(pheno_data$path), pData=pheno_data)\n",
    "\n",
    "# Display a description of this object\n",
    "bg\n",
    "\n",
    "# Load all attributes including gene name\n",
    "bg_table = texpr(bg, 'all')\n",
    "bg_gene_names = unique(bg_table[, 9:10])\n",
    "\n",
    "# Save the ballgown object to a file for later use\n",
    "save(bg, file='/home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/bg.rda')\n",
    "\n",
    "# Perform differential expression (DE) analysis with no filtering\n",
    "results_transcripts = stattest(bg, feature=\"transcript\", covariate=\"type\", getFC=TRUE, meas=\"FPKM\")\n",
    "results_genes = stattest(bg, feature=\"gene\", covariate=\"type\", getFC=TRUE, meas=\"FPKM\")\n",
    "results_genes = merge(results_genes, bg_gene_names, by.x=c(\"id\"), by.y=c(\"gene_id\"))\n",
    "\n",
    "# Save a tab delimited file for both the transcript and gene results\n",
    "write.table(results_transcripts, \"/home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/Control_vs_PDX_transcript_results.tsv\", sep=\"\\t\", quote=FALSE, row.names = FALSE)\n",
    "write.table(results_genes, \"/home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/Control_vs_PDX_gene_results.tsv\", sep=\"\\t\", quote=FALSE, row.names = FALSE)\n",
    "\n",
    "# Filter low-abundance genes. Here we remove all transcripts with a variance across the samples of less than one\n",
    "bg_filt = subset (bg,\"rowVars(texpr(bg)) > 1\", genomesubset=TRUE)\n",
    "\n",
    "# Load all attributes including gene name\n",
    "bg_filt_table = texpr(bg_filt , 'all')\n",
    "bg_filt_gene_names = unique(bg_filt_table[, 9:10])\n",
    "\n",
    "# Perform DE analysis now using the filtered data\n",
    "results_transcripts = stattest(bg_filt, feature=\"transcript\", covariate=\"type\", getFC=TRUE, meas=\"FPKM\")\n",
    "results_genes = stattest(bg_filt, feature=\"gene\", covariate=\"type\", getFC=TRUE, meas=\"FPKM\")\n",
    "results_genes = merge(results_genes, bg_filt_gene_names, by.x=c(\"id\"), by.y=c(\"gene_id\"))\n",
    "\n",
    "# Output the filtered list of genes and transcripts and save to tab delimited files\n",
    "write.table(results_transcripts, \"/home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/Control_vs_PDX_transcript_results_filtered.tsv\", sep=\"\\t\", quote=FALSE, row.names = FALSE)\n",
    "write.table(results_genes, \"/home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/Control_vs_PDX_gene_results_filtered.tsv\", sep=\"\\t\", quote=FALSE, row.names = FALSE)\n",
    "\n",
    "# Identify the significant genes with p-value < 0.05\n",
    "sig_transcripts = subset(results_transcripts, results_transcripts$pval<0.05)\n",
    "sig_genes = subset(results_genes, results_genes$pval<0.05)\n",
    "\n",
    "# Output the significant gene results to a pair of tab delimited files\n",
    "write.table(sig_transcripts, \"/home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/Control_vs_PDX_transcript_results_sig.tsv\", sep=\"\\t\", quote=FALSE, row.names = FALSE)\n",
    "write.table(sig_genes, \"/home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/Control_vs_PDX_gene_results_sig.tsv\", sep=\"\\t\", quote=FALSE, row.names = FALSE)\n",
    "\n",
    "# Exit the R session\n",
    "quit(save=\"no\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3dc5e12-144e-439f-a887-0367378007df",
   "metadata": {},
   "source": [
    "View the raw output from Ballgown.\n",
    "\n",
    "```head /home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/Control_vs_PDX_gene_results.tsv | column -t```\n",
    "\n",
    "View how many genes are present on this chromosome.\n",
    "\n",
    "```grep -v feature /home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/Control_vs_PDX_gene_results.tsv | wc -l```\n",
    "\n",
    "Output: 55487\n",
    "\n",
    "View the genes that passed the filtering step.\n",
    "\n",
    "```grep -v feature /home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/Control_vs_PDX_gene_results_filtered.tsv | wc -l```\n",
    "\n",
    "Output: 5301\n",
    "\n",
    "View how many genes were considered differentially expressed (p < 0.05).\n",
    "\n",
    "```grep -v feature /home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/Control_vs_PDX_gene_results_sig.tsv | wc -l```\n",
    "\n",
    "Output: 1254\n",
    "\n",
    "As an exercise or just a visualization, you can look at the top 20 DE genes in IGV and see if they seem to make sense.\n",
    "\n",
    "```grep -v feature /home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/Control_vs_PDX_gene_results_sig.tsv | sort -rnk 3 | head -n 20 | column -t``` higher abundance in CON?\n",
    "\n",
    "```grep -v feature /home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/Control_vs_PDX_gene_results_sig.tsv | sort -nk 3 | head -n 20 | column -t``` higher abundance in PDX?\n",
    "\n",
    "Save all the genes that have a p < 0.05.\n",
    "\n",
    "```grep -v feature /home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/Control_vs_PDX_gene_results_sig.tsv | cut -f 6 | sed 's/\\\"//g' > /home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/DE_genes.txt```\n",
    "\n",
    "```head /home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/DE_genes.txt```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e214b8d2-edb7-48a0-a457-1d86cfe98026",
   "metadata": {},
   "source": [
    "***\n",
    "# Ballgown DE Visualization\n",
    "\n",
    "```cd /home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only```\n",
    "\n",
    "Run the R code in the cell below. Uncomment all the lines with ctrl + / if needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71639197-7575-4bd2-8797-c5893f5b5086",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load libraries\n",
    "library(ballgown)\n",
    "library(genefilter)\n",
    "library(dplyr)\n",
    "library(devtools)\n",
    "\n",
    "#designate output file\n",
    "outfile=\"/home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/BP_PRJNA496042_ballgown_output.pdf\"\n",
    "\n",
    "# Generate phenotype data\n",
    "ids = c(\"Control_1\",\"Control_2\",\"Control_3\",\"Control_4\",\"PDX_1\",\"PDX_2\",\"PDX_3\",\"PDX_4\")\n",
    "type = c(\"Control\",\"Control\",\"Control\",\"Control\",\"PDX\",\"PDX\",\"PDX\",\"PDX\")\n",
    "results = \"/home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/\" # be sure to include the trailing \"/\"\n",
    "path = paste(results,ids,sep=\"\")\n",
    "pheno_data = data.frame(ids,type,path)\n",
    "\n",
    "# Display the phenotype data\n",
    "pheno_data\n",
    "\n",
    "# Load the ballgown object from file\n",
    "# Make sure to set the path to 'bg.rda', this file was created in the ballgown DE R script\n",
    "load('/home/user/bioinformatics/BP_PRJNA496042/alignments/de/ballgown/ref_only/bg.rda')\n",
    "\n",
    "# The load command, loads an R object from a file into memory in our R session.\n",
    "# You can use ls() to view the names of variables that have been loaded\n",
    "ls()\n",
    "\n",
    "# Print a summary of the ballgown object\n",
    "bg\n",
    "\n",
    "# Open a PDF file where we will save some plots. We will save all figures and then view the PDF at the end\n",
    "pdf(file=outfile)\n",
    "\n",
    "# Extract FPKM values from the 'bg' object\n",
    "fpkm = texpr(bg,meas=\"FPKM\")\n",
    "\n",
    "# View the last several rows of the FPKM table\n",
    "tail(fpkm)\n",
    "\n",
    "# Transform the FPKM values by adding 1 and convert to a log2 scale\n",
    "# Adding 1 to the FPKM is known as Laplace smoothing, it prevents issues with taking the log of a \n",
    "# zero or near-zero value by ensuring it is always positive.\n",
    "# The log2 transform compresses the range of values, allowing easier visualization.\n",
    "fpkm = log2(fpkm+1)\n",
    "\n",
    "# View the last several rows of the transformed FPKM table\n",
    "tail(fpkm)\n",
    "\n",
    "# Create boxplots to display summary statistics for the FPKM values for each sample\n",
    "boxplot(fpkm,col=as.numeric(as.factor(pheno_data$type))+1,las=2,ylab='log2(FPKM+1)')\n",
    "\n",
    "# col=as.numeric(as.factor(pheno_data$type))+1 - set color based on pheno_data$type which is Control vs PDX\n",
    "# las=2 - labels are perpendicular to axis \n",
    "# ylab='log2(FPKM+1)' - set ylab to indicate that values are log2 transformed\n",
    "\n",
    "# Display the transcript ID for a single row of data\n",
    "ballgown::transcriptNames(bg)[2763]\n",
    "\n",
    "# Display the gene name for a single row of data\n",
    "ballgown::geneNames(bg)[2763]\n",
    "\n",
    "# Create a BoxPlot comparing the expression of a single gene for all replicates of both conditions\n",
    "boxplot(fpkm[2763,] ~ pheno_data$type, border=c(2,3), main=paste(ballgown::geneNames(bg)[2763],' : ', ballgown::transcriptNames(bg)[2763]),pch=19, xlab=\"Type\", ylab='log2(FPKM+1)')\n",
    "\n",
    "# border=c(2,3) - set border color for each of the boxplots\n",
    "# main=paste(ballgown::geneNames(bg)[2763],' : ', ballgown::transcriptNames(bg)[2763]) - set title to gene : transcript\n",
    "# xlab=\"Type\" - set x label to Type\n",
    "# ylab='log2(FPKM+1)' - set ylab to indicate that values are log2 transformed\n",
    "\n",
    "\n",
    "# Add the FPKM values for each sample onto the plot\n",
    "points(fpkm[2763,] ~ jitter(c(2,2,2,1,1,1)), col=c(2,2,2,1,1,1)+1, pch=16)\n",
    "# pch=16 - set plot symbol to solid circle, default is empty circle\n",
    "\n",
    "\n",
    "# Create a plot of transcript structures observed in each replicate and color transcripts by expression level\n",
    "plotTranscripts(ballgown::geneIDs(bg)[2763], bg, main=c('TST in all HBR samples'), sample=c('HBR_Rep1', 'HBR_Rep2', 'HBR_Rep3'), labelTranscripts=TRUE)\n",
    "plotTranscripts(ballgown::geneIDs(bg)[2763], bg, main=c('TST in all UHR samples'), sample=c('UHR_Rep1', 'UHR_Rep2', 'UHR_Rep3'), labelTranscripts=TRUE)\n",
    "\n",
    "#plotMeans('TST',bg,groupvar=\"type\",legend=FALSE)\n",
    "\n",
    "# Close the PDF device where we have been saving our plots\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
