{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2e9aeba0-1800-4065-991b-0676770caea6",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Background"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4bb275d-0e96-4f34-bb5a-cd2eb040cbda",
   "metadata": {},
   "source": [
    "This RNA-seq workflow is adapted from the RNA-seq tutuorial at [rnabio.org](https://rnabio.org).\n",
    "\n",
    "``` Malachi Griffith*, Jason R. Walker, Nicholas C. Spies, Benjamin J. Ainscough, Obi L. Griffith*. 2015. Informatics for RNA-seq: A web resource for analysis on the cloud. PLoS Comp Biol. 11(8):e1004393```\n",
    "\n",
    "The [ENCODE project](https://www.encodeproject.org) publishes [guidelines and best practices for RNA-seq](https://www.encodeproject.org/documents/cede0cbe-d324-4ce7-ace4-f0c3eddf5972/@@download/attachment/ENCODE%20Best%20Practices%20for%20RNA_v2.pdf) that are a must read."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99c94222-8085-4976-b73a-b5c5a2a11eaa",
   "metadata": {},
   "source": [
    "## Initial setup\n",
    "The tutorial being followed uses AWS cloud but I am writing this pipeline to be ran locally on a linux-based machine. More specifically, I am using Manjaro with kernel version 6.1.7-1-MANJARO (64-bit) and KDE plasma version 5.26.4 on X11.\n",
    "\n",
    "[Configure your environment](https://rnabio.org/module-00-setup/0000/09/01/Environment/) and then [install all necessary software](https://rnabio.org/module-00-setup/0000/10/01/Installation/) according to the rnabio website. The specific steps you follow while doing this will depend on your system and how you want to install the tools. I would suggest looking for maintained packages for your specific distribution before trying to install from source.\n",
    "\n",
    "Specific verison information of software installed on my setup:\n",
    "- SAMtools v1.16.1, installed from AUR\n",
    "    - run with ```samtools``` from CL\n",
    "- bam-readcount v1.0.1, compiled from source\n",
    "    - run from ```~$HOME/bioinformatics/bam-readcount/build/bin``` with ```./bam-readcount```\n",
    "- HISAT2 v2.2.1-1, installed from AUR\n",
    "    - run with ```hisat2``` from CL\n",
    "- stringtie v2.2.1, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/stringtie-2.2.1.Linux_x86_64``` with ```./stringtie```\n",
    "- gffcompare v0.12.6-2, installed from AUR\n",
    "    - run with ```gffcompare``` from CL\n",
    "- htseq-count v.0.11.3-1, installed from AUR and also in bioinformatics conda env\n",
    "    - run with ```htseq-count``` from terminal with conda bioinformatics env active\n",
    "- tophat v2.1.1, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/tophat-2.1.1.Linux_x86_64``` with ```./gtf_to_fasta```\n",
    "- kallisto v0.48.0-1, installed from AUR\n",
    "    - run with ```kallisto``` from CL\n",
    "- fastqc v0.11.9-1, installed from AUR\n",
    "    - run with ```fastqc``` from CL\n",
    "- multiqc v1.14, installed with pip3 in bioinformatics conda env\n",
    "    - run with ```multiqc --help``` from CL\n",
    "- picard v2.26.4, downloaded as JAR file\n",
    "    - run from ```~$HOME/bioinformatics/``` with ```java -jar picard.jar```\n",
    "- flexbar v3.5.0, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/flexbar-3.5.0-linux``` with ```./flexbar```\n",
    "- regtools v1.0.0, installed from AUR (git source)\n",
    "    - run with ```regtools``` from CL\n",
    "- rseqc v5.0.1, installed with pip3 in bioinformatics conda env\n",
    "    - import in python with ```rseqc``` or run with ```read_GC.py``` in CL?\n",
    "- bedops v2.4.40, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/bedops_linux_x86_64-v2.4.40/bin``` with```./bedops``` and ```./gff2bed```\n",
    "- gtfToGenePred v unknown, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/gtfToGenePred``` with ```./gtfToGenePred```\n",
    "- genePredToBed v unknown, downloaded pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/genePredToBed``` with ```./genePredToBed```\n",
    "- how_are_we_stranded_here, v unknown, installed with pip3 in bioinformatics conda env\n",
    "    - run with ```check_strandedness``` in CL\n",
    "- cell ranger v7.1.0, have to register to download, pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics/cellranger-7.1.0``` with ```./cellranger```\n",
    "- tabix, website only offers apt installation, not relevant on arch-based distro. Tabix seems to be included with HTSlib which is already installed as a dependency for samtools and bcftools\n",
    "- BWA v0.7.17=r1188, compiled from source\n",
    "    - run from ```~$HOME/bioinformatics/bwa``` with ```./bwa``` or ```bwa```\n",
    "- BCFtools v1.16, installed from AUR\n",
    "    - run with ```bcftools``` from CL\n",
    "- peddy v0.4.8, cloned and installed with pip in bioinformatics conda env\n",
    "    - run with ```peddy``` from CL in conda env\n",
    "- slivar v0.2.7, pre-complied binary\n",
    "    - run from ```~$HOME/bioinformatics``` with ```./slivar```\n",
    "- STRling v0.5.1, pre-complied binary\n",
    "    - run from ```~$HOME/bioinformatics``` with ```./strling```\n",
    "- freebayes v1.3.6, pre-compiled binary\n",
    "    - run from ```~$HOME/bioinformatics``` with ```./freebayes-1.3.6-linux-amd64-static```\n",
    "\n",
    "### R Libraries needed:\n",
    "- devtools\n",
    "- dplyr\n",
    "- gplots\n",
    "- ggplot2\n",
    "\n",
    "install with \n",
    "```install.packages(c(\"devtools\",\"dplyr\",\"gplots\",\"ggplot2\"),repos=\"http://cran.us.r-project.org\")```\n",
    "\n",
    "### Bioconductor libraries needed:\n",
    "- [genefilter](http://bioconductor.org/packages/release/bioc/html/genefilter.html)\n",
    "- [ballgown](http://bioconductor.org/packages/release/bioc/html/ballgown.html)\n",
    "- [edgeR](http://www.bioconductor.org/packages/release/bioc/html/edgeR.html)\n",
    "- [GenomicRanges](http://bioconductor.org/packages/release/bioc/html/GenomicRanges.html)\n",
    "- [rhdf5](https://www.bioconductor.org/packages/release/bioc/html/rhdf5.html)\n",
    "- [biomaRt](https://bioconductor.org/packages/release/bioc/html/biomaRt.html)\n",
    "\n",
    "### Sleuth\n",
    "[Source](https://pachterlab.github.io/sleuth/download)\n",
    "\n",
    "devtools need to be installed if not already ```install.packages(\"devtools\")```\n",
    "\n",
    "install with ```devtools::install_github(\"pachterlab/sleuth\")```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1afe64c2-2756-44ba-9b0b-e9189f2d2313",
   "metadata": {},
   "source": [
    "***\n",
    "# General notes on RNA-seq workflow\n",
    "\n",
    "Due to inconsistencies in the naming conventions for chromosomes, you should obtain your reference genome and annotation packages from the same source (Ensembl, NCBI, UCSC). Additionally, your annotations must correspond with the same reference genome version (i.e. both correspond to NCBI human GRCh38).\n",
    "\n",
    "[This guide](https://pmbio.org/module-02-inputs/0002/02/01/Reference_Genome/) explains reference genomes and how to obtain them.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff56a121-6803-4e44-9c4a-0fafa432329d",
   "metadata": {},
   "source": [
    "***\n",
    "# Obtain known gene/transcript annotations\n",
    "\n",
    "Download the annotations from Ensembl via ```wget http://genomedata.org/rnaseq-tutorial/annotations/GRCh38/chr22_with_ERCC92.gtf```. \n",
    "\n",
    "You can preview some of the file with ```cat chr22_with_ERCC92.gtf | column -t | less -p exon -S```.\n",
    "\n",
    "\n",
    "```cat chr22_with_ERCC92.gtf | grep -w gene | wc -l```\n",
    "\n",
    "You can view the structure of one transcript with ```grep ENST00000342247 $RNA_REF_GTF | less -p \"exon\\s\" -S```. Exit with q.\n",
    "\n",
    "Other sources of annontation files for a HISAT2/StringTie/Ballgown pipeline, refer to [here](https://rnabio.org/module-01-inputs/0001/03/01/Annotations/).\n",
    "\n",
    "HISAT2 has prebuilt indexes for DNA and RNA for both Ensembl and UCSC genomes [here](https://ccb.jhu.edu/software/hisat2/index.shtml).\n",
    "\n",
    "**Important**: chromosome names in the .gtf file must be the same as those in the reference genome .fa file. If StringTie returns expression values that are all 0, this is the likley cause of that error. Download the annotation and reference genome files from the same source.\n",
    "\n",
    "**Make sure your reference genome and annotations are for the same build of the genome (i.e. both for GRCh38).**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f10eeb8-390b-4706-b4b4-6650f0ce9940",
   "metadata": {},
   "source": [
    "***\n",
    "# Creating index with HISAT2\n",
    "\n",
    "This was done directly using python files from the [HISAT2 github repo](https://github.com/DaehwanKimLab/hisat2), not sure how to do it using the installed program listed above or if it's possible. Just clone the repo and use the files necessary.\n",
    "\n",
    "Run the following commands from the location of your reference .gtf and .fa files. The chr22_with_ERCC92.gtf and chr22_with_ERCC92.fa. You may need a ```./``` in front of the first two commands\n",
    "\n",
    "```hisat2_extract_splice_sites.py chr22_with_ERCC92.gtf > splicesites.tsv```\n",
    "\n",
    "```hisat2_extract_exons.py chr22_with_ERCC92.gtf >exons.tsv```\n",
    "\n",
    "```hisat2-build -p 8 --ss splicesites.tsv --exon exons.tsv chr22_with_ERCC92.fa chr22_with_ERCC92```\n",
    "\n",
    "The output should be a collection of binary files with names like chr22_with_ERCC92.x.ht2, the x being a number."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b06fc2fd-5eff-424d-b4a7-296c410a9ce9",
   "metadata": {},
   "source": [
    "***\n",
    "# RNA-seq test data\n",
    "\n",
    "```mkdir test_data && cd test_data```\n",
    "\n",
    "```wget http://genomedata.org/rnaseq-tutorial/HBR_UHR_ERCC_ds_5pc.tar```\n",
    "\n",
    "unpack the tarball\n",
    "\n",
    "```tar -xvf HBR_UHR_ERCC_ds_5pc.tar```\n",
    "\n",
    "Look at the head of one of the files.\n",
    "\n",
    "```zcat UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz | head -n 8```\n",
    "\n",
    "Determine how many reads are in the first library.\n",
    "\n",
    "```zcat UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz | grep -P \"^\\@HWI\" | wc -l```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "796edb8a-4e09-4ba4-812c-8d06e5bc8df3",
   "metadata": {},
   "source": [
    "***\n",
    "# Check strandedness (optional, but not really)\n",
    "\n",
    "This step is not required but due to differences in usage of RNA-seq library construction kits, it is **highly** recommended to empirically verify strandedness using the below method.\n",
    "\n",
    "The check_strandedness should already be installed at the beginning. It can also be installed as a [Docker image](https://hub.docker.com/r/mgibio/checkstrandedness).\n",
    "\n",
    "Convert gtf to genepred\n",
    "\n",
    "```gtfToGenePred chr22_with_ERCC92.gtf chr22_with_ERCC92.genePred```\n",
    "\n",
    "Convert genPred to bed12\n",
    "\n",
    "```genePredToBed chr22_with_ERCC92.genePred chr22_with_ERCC92.bed12```\n",
    "\n",
    "Use bedtools to create fasta from GTF\n",
    "\n",
    "```bedtools getfasta -fi chr22_with_ERCC92.fa -bed chr22_with_ERCC92.bed12 -s -split -name -fo chr22_ERCC92.transcripts.fa```\n",
    "\n",
    "Clean up the header lines of the transcript file\n",
    "\n",
    "```cat chr22_ERCC92_transcripts.fa | perl -ne 'if($_ =~/^\\>\\S+\\:\\:(ERCC\\-\\d+)\\:.*/){print \">$1\\n\"}elsif ($_ =~/^\\>(\\S+)\\:\\:.*/){print \">$1\\n\"}else{print $_}' > chr22_ERCC92_transcripts.clean.fa```\n",
    "\n",
    "View the clean file\n",
    "\n",
    "```less chr22_ERCC92_transcripts.clean.fa```\n",
    "\n",
    "Add in transcript_id field to .gtf file\n",
    "\n",
    "```awk '{ if ($0 ~ \"transcript_id\") print $0; else print $0\" transcript_id \\\"\\\";\"; }' chr22_with_ERCC92.gtf > chr22_with_ERCC92_tidy.gtf```\n",
    "\n",
    "check strandedness \n",
    "\n",
    "```check_strandedness --gtf chr22_with_ERCC92_tidy.gtf --transcripts chr22_ERCC92_transcripts.clean.fa --reads_1 test_data/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz --reads_2 test_data/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz```\n",
    "\n",
    "OUTPUT:\n",
    "\n",
    "Fraction of reads failed to determine: 0.1123\n",
    "\n",
    "Fraction of reads explained by \"1++,1--,2+-,2-+\": 0.0155 (1.7% of explainable reads)\n",
    "\n",
    "Fraction of reads explained by \"1+-,1-+,2++,2--\": 0.8722 (98.3% of explainable reads)\n",
    "\n",
    "Over 90% of reads explained by \"1+-,1-+,2++,2--\"\n",
    "\n",
    "Data is likely RF/fr-firststrand\n",
    "\n",
    "More info on strandedness can be found [here](https://rnabio.org/module-09-appendix/0009/12/01/StrandSettings/)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "764cd5a0-fa7a-4dc1-9f67-335e69350306",
   "metadata": {},
   "source": [
    "***\n",
    "# Pre-alignment QC\n",
    "\n",
    "To get a sense of the raw reads quality in your fastq files, run the following programs.\n",
    "\n",
    "```fastqc *.fastq.gz``` in the same directory as your reads.\n",
    "\n",
    "It will output an HTML file for each read in the same directory that contains a full report of the quality measures.\n",
    "\n",
    "In the \"over-represented sequences\" section, try [BLASTing](https://blast.ncbi.nlm.nih.gov/Blast.cgi) the sequences to try determining a source organism.\n",
    "\n",
    "Run multiqc to combine all the fastqc reports.\n",
    "\n",
    "```multiqc .``` run in the same directory as the reports.\n",
    "\n",
    "Make a new directory for the fastqc .html outputs and move them\n",
    "\n",
    "```mkdir fastqc && mv *_fastqc* fastqc```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b9099f7-50d6-42da-91bc-2853d7b6a3bf",
   "metadata": {},
   "source": [
    "***\n",
    "# Trim adapters\n",
    "\n",
    "Make a new folder for the trimmed reads\n",
    "\n",
    "```mkdir trimmed```\n",
    "\n",
    "Download the Illumina adapter sequence files, or the relevant files for your sequencing machine.\n",
    "\n",
    "```wget http://genomedata.org/rnaseq-tutorial/illumina_multiplex.fa```\n",
    "\n",
    "Use flexbar to remove adapter sequences and trim first 13 bases of each read. Adjust the number of threads to something manageable for your system. It is set to 24 threads and takes a few seconds to run for each command. It may take much longer with a slower CPU.\n",
    "\n",
    "Possible options for flexbar and their meaning:\n",
    "\n",
    "   - ```–adapter-min-overlap 7``` requires a minimum of 7 bases to match the adapter\n",
    "   - ```–adapter-trim-end RIGHT``` uses a trimming strategy to remove the adapter from the 3 prime or RIGHT end of the read\n",
    "   - ```–max-uncalled 300``` allows as many as 300 uncalled or N bases (MiSeq read lengths can be 300bp)\n",
    "   - ```–min-read-length``` the minimum read length allowed after trimming is 25bp.\n",
    "   - ```–threads 8``` use 8 threads\n",
    "   - ```–zip-output GZ``` the input FASTQ files are gzipped so we will output gzipped FASTQ to save space\n",
    "   - ```–adapters``` define the path to the adapter FASTA file to trim\n",
    "   - ```–reads``` define the path to the read 1 FASTQ file of reads\n",
    "   - ```–reads2``` define the path to the read 2 FASTQ file of reads\n",
    "   - ```–target``` a base path for the output files. The value will _1.fastq.gz and _2.fastq.gz for read 1 and read 2 respectively\n",
    "   - ```–pre-trim-left``` trim a fixed number of bases at left read end. For example, to trim 5 bases at the left side of reads: –pre-trim-left 5\n",
    "   - ```–pre-trim-right``` trim a fixed number of bases at right read end. For example, to trim 5 bases at the right side of reads: –pre-trim-right 5\n",
    "   - ```–pre-trim-phred``` trim based on phred quality value to deal with higher error rates towards the end of reads. For example, to trim the 3 prime end until quality offset value 30 or higher is reached, specify: –pre-trim-phred 30\n",
    "\n",
    "The commands to run for flexbar for this dataset are as follows:\n",
    "\n",
    "```./flexbar --adapter-min-overlap 7 --adapter-trim-end RIGHT --adapters /home/user/bioinformatics/ref_genome/illumina_multiplex.fa --pre-trim-left 13 --max-uncalled 300 --min-read-length 25 --threads 24 --zip-output GZ --reads /home/user/bioinformatics/ref_genome/test_data/UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz --reads2 /home/user/bioinformatics/ref_genome/test_data/UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz --target /home/user/bioinformatics/ref_genome/test_data/trimmed/UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22```\n",
    "\n",
    "```./flexbar --adapter-min-overlap 7 --adapter-trim-end RIGHT --adapters /home/user/bioinformatics/ref_genome/illumina_multiplex.fa --pre-trim-left 13 --max-uncalled 300 --min-read-length 25 --threads 24 --zip-output GZ --reads /home/user/bioinformatics/ref_genome/test_data/UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz --reads2 /home/user/bioinformatics/ref_genome/test_data/UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz --target /home/user/bioinformatics/ref_genome/test_data/trimmed/UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22```\n",
    "\n",
    "```./flexbar --adapter-min-overlap 7 --adapter-trim-end RIGHT --adapters /home/user/bioinformatics/ref_genome/illumina_multiplex.fa --pre-trim-left 13 --max-uncalled 300 --min-read-length 25 --threads 24 --zip-output GZ --reads /home/user/bioinformatics/ref_genome/test_data/UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz --reads2 /home/user/bioinformatics/ref_genome/test_data/UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz --target /home/user/bioinformatics/ref_genome/test_data/trimmed/UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22```\n",
    "\n",
    "```./flexbar --adapter-min-overlap 7 --adapter-trim-end RIGHT --adapters /home/user/bioinformatics/ref_genome/illumina_multiplex.fa --pre-trim-left 13 --max-uncalled 300 --min-read-length 25 --threads 24 --zip-output GZ --reads /home/user/bioinformatics/ref_genome/test_data/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz --reads2 /home/user/bioinformatics/ref_genome/test_data/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz --target /home/user/bioinformatics/ref_genome/test_data/trimmed/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22```\n",
    "\n",
    "```./flexbar --adapter-min-overlap 7 --adapter-trim-end RIGHT --adapters /home/user/bioinformatics/ref_genome/illumina_multiplex.fa --pre-trim-left 13 --max-uncalled 300 --min-read-length 25 --threads 24 --zip-output GZ --reads /home/user/bioinformatics/ref_genome/test_data/HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz --reads2 /home/user/bioinformatics/ref_genome/test_data/HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz --target /home/user/bioinformatics/ref_genome/test_data/trimmed/HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22```\n",
    "\n",
    "```./flexbar --adapter-min-overlap 7 --adapter-trim-end RIGHT --adapters /home/user/bioinformatics/ref_genome/illumina_multiplex.fa --pre-trim-left 13 --max-uncalled 300 --min-read-length 25 --threads 24 --zip-output GZ --reads /home/user/bioinformatics/ref_genome/test_data/HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz --reads2 /home/user/bioinformatics/ref_genome/test_data/HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz --target /home/user/bioinformatics/ref_genome/test_data/trimmed/HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22```\n",
    "\n",
    "Compare the impact of trimming with the original QC reports by running fastqc and multiqc on the trimmed reads\n",
    "\n",
    "In the directory of the trimmed reads, run:\n",
    "\n",
    "```fastqc *.fastq.gz```\n",
    "\n",
    "```multiqc .```\n",
    "\n",
    "Same as before, make a new directory for the fastqc .html outputs and move them\n",
    "\n",
    "```mkdir fastqc && mv *_fastqc* fastqc```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf703eb9-5218-4e71-80c4-8b4e376b6eff",
   "metadata": {},
   "source": [
    "***\n",
    "# Alignment with HISAT2\n",
    "\n",
    "For more details on HISAT2, refer to the [manual on Github](http://daehwankimlab.github.io/hisat2/manual/)\n",
    "\n",
    "The outputs after running HISAT2 will be a SAM/BAM file for each data set.\n",
    "\n",
    "Options for HISAT2 include:\n",
    "\n",
    "\n",
    "   - ```-p 8``` tells HISAT2 to use eight threads for bowtie alignments.\n",
    "   - ```–rna-strandness RF``` specifies strandness of RNAseq library. We will specify RF since the TruSeq strand-specific library was used to make these libraries. See here for options.\n",
    "   - ```–rg-id $ID``` specifies a read group ID that is a unique identifier.\n",
    "   - ```–rg SM:$SAMPLE_NAME``` specifies a read group sample name. This together with rg-id will allow you to determine which reads came from which sample in the merged bam later on.\n",
    "   - ```–rg LB:$LIBRARY_NAME``` specifies a read group library name. This together with rg-id will allow you to determine which reads came from which library in the merged bam later on.\n",
    "   - ```–rg PL:ILLUMINA``` specifies a read group sequencing platform.\n",
    "   - ```–rg PU:$PLATFORM_UNIT``` specifies a read group sequencing platform unit. Typically this consists of FLOWCELL-BARCODE.LANE\n",
    "   - ```–dta``` Reports alignments tailored for transcript assemblers.\n",
    "   - ```-x /path/to/hisat2/index``` The HISAT2 index filename prefix (minus the trailing .X.ht2) built earlier including splice sites and exons.\n",
    "   - ```-1 /path/to/read1.fastq.gz``` The read 1 FASTQ file, optionally gzip(.gz) or bzip2(.bz2) compressed.\n",
    "   - ```-2 /path/to/read2.fastq.gz``` The read 2 FASTQ file, optionally gzip(.gz) or bzip2(.bz2) compressed.\n",
    "   - ```-S /path/to/output.sam``` The output SAM format text file of alignments.\n",
    "\n",
    "Run a command for each of your replicates.\n",
    "\n",
    "```hisat2 -p 24 --rg-id=UHR_Rep1 --rg SM:UHR --rg LB:UHR_Rep1_ERCC-Mix1 --rg PL:ILLUMINA --rg PU:CXX1234-ACTGAC.1 -x /home/user/bioinformatics/ref_genome/chr22_with_ERCC92 --dta --rna-strandness RF -1 /home/user/bioinformatics/ref_genome/test_data/UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz -2 /home/user/bioinformatics/ref_genome/test_data/UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz -S /home/user/bioinformatics/ref_genome/test_data/align_output/UHR_Rep1.sam```\n",
    "\n",
    "```hisat2 -p 24 --rg-id=UHR_Rep2 --rg SM:UHR --rg LB:UHR_Rep2_ERCC-Mix1 --rg PL:ILLUMINA --rg PU:CXX1234-TGACAC.1 -x /home/user/bioinformatics/ref_genome/chr22_with_ERCC92 --dta --rna-strandness RF -1 /home/user/bioinformatics/ref_genome/test_data/UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz -2 /home/user/bioinformatics/ref_genome/test_data/UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz -S /home/user/bioinformatics/ref_genome/test_data/align_output/UHR_Rep2.sam```\n",
    "\n",
    "```hisat2 -p 24 --rg-id=UHR_Rep3 --rg SM:UHR --rg LB:UHR_Rep3_ERCC-Mix1 --rg PL:ILLUMINA --rg PU:CXX1234-CTGACA.1 -x /home/user/bioinformatics/ref_genome/chr22_with_ERCC92 --dta --rna-strandness RF -1 /home/user/bioinformatics/ref_genome/test_data/UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz -2 /home/user/bioinformatics/ref_genome/test_data/UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz -S /home/user/bioinformatics/ref_genome/test_data/align_output/UHR_Rep3.sam```\n",
    "\n",
    "```hisat2 -p 24 --rg-id=HBR_Rep1 --rg SM:HBR --rg LB:HBR_Rep1_ERCC-Mix2 --rg PL:ILLUMINA --rg PU:CXX1234-TGACAC.1 -x /home/user/bioinformatics/ref_genome/chr22_with_ERCC92 --dta --rna-strandness RF -1 /home/user/bioinformatics/ref_genome/test_data/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz -2 /home/user/bioinformatics/ref_genome/test_data/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz -S /home/user/bioinformatics/ref_genome/test_data/align_output/HBR_Rep1.sam```\n",
    "\n",
    "```hisat2 -p 24 --rg-id=HBR_Rep2 --rg SM:HBR --rg LB:HBR_Rep2_ERCC-Mix2 --rg PL:ILLUMINA --rg PU:CXX1234-GACACT.1 -x /home/user/bioinformatics/ref_genome/chr22_with_ERCC92 --dta --rna-strandness RF -1 /home/user/bioinformatics/ref_genome/test_data/HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz -2 /home/user/bioinformatics/ref_genome/test_data/HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz -S /home/user/bioinformatics/ref_genome/test_data/align_output/HBR_Rep2.sam```\n",
    "\n",
    "```hisat2 -p 24 --rg-id=HBR_Rep3 --rg SM:HBR --rg LB:HBR_Rep3_ERCC-Mix2 --rg PL:ILLUMINA --rg PU:CXX1234-ACACTG.1 -x /home/user/bioinformatics/ref_genome/chr22_with_ERCC92 --dta --rna-strandness RF -1 /home/user/bioinformatics/ref_genome/test_data/HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz -2 /home/user/bioinformatics/ref_genome/test_data/HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz -S /home/user/bioinformatics/ref_genome/test_data/align_output/HBR_Rep3.sam```\n",
    "\n",
    "### Convert SAM to BAM files\n",
    "\n",
    "```samtools sort -@ 8 -o UHR_Rep1.bam UHR_Rep1.sam```\n",
    "```samtools sort -@ 8 -o UHR_Rep2.bam UHR_Rep2.sam```\n",
    "```samtools sort -@ 8 -o UHR_Rep3.bam UHR_Rep3.sam```\n",
    "```samtools sort -@ 8 -o HBR_Rep1.bam HBR_Rep1.sam```\n",
    "```samtools sort -@ 8 -o HBR_Rep2.bam HBR_Rep2.sam```\n",
    "```samtools sort -@ 8 -o HBR_Rep3.bam HBR_Rep3.sam```\n",
    "\n",
    "### Merge all BAM files\n",
    "Other tools such as \"samtools merge\" and \"bamtools merge\" can also achieve this. Create a BAM file for each group (UHR and HBR).\n",
    "\n",
    "```java -Xmx2g -jar picard.jar MergeSamFiles -OUTPUT /home/user/bioinformatics/ref_genome/test_data/align_output/UHR.bam -INPUT /home/user/bioinformatics/ref_genome/test_data/align_output/UHR_Rep1.bam -INPUT /home/user/bioinformatics/ref_genome/test_data/align_output/UHR_Rep2.bam -INPUT /home/user/bioinformatics/ref_genome/test_data/align_output/UHR_Rep3.bam```\n",
    "\n",
    "```java -Xmx2g -jar picard.jar MergeSamFiles -OUTPUT /home/user/bioinformatics/ref_genome/test_data/align_output/HBR.bam -INPUT /home/user/bioinformatics/ref_genome/test_data/align_output/HBR_Rep1.bam -INPUT /home/user/bioinformatics/ref_genome/test_data/align_output/HBR_Rep2.bam -INPUT /home/user/bioinformatics/ref_genome/test_data/align_output/HBR_Rep3.bam```\n",
    "\n",
    "You can verify that all bam files were created successfully, there should be 8.\n",
    "\n",
    "```ls -l *.bam | wc -l```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef451cfe-f5b2-4877-bf11-a52fdc18eaec",
   "metadata": {},
   "source": [
    "***\n",
    "# Using the Integrative Genomics Viewer (IGV)\n",
    "\n",
    "IGV can be run from the download directory using ```./igv.sh```\n",
    "\n",
    "To re-visit the shortish tutorial on navigating IGV, go [here](https://rnabio.org/module-02-alignment/0002/04/01/IGV/). You can follow the tutorial there using the provided files.\n",
    "\n",
    "For our dataset, we need to index the BAM files before loading them into IGV. This will be done with samtools. From the directory where the BAM files are located, run:\n",
    "\n",
    "```samtools index HBR.bam```\n",
    "\n",
    "```samtools index HBR_Rep1.bam```\n",
    "\n",
    "```samtools index HBR_Rep2.bam```\n",
    "\n",
    "```samtools index HBR_Rep3.bam```\n",
    "\n",
    "```samtools index UHR.bam```\n",
    "\n",
    "```samtools index UHR_Rep1.bam```\n",
    "\n",
    "```samtools index UHR_Rep2.bam```\n",
    "\n",
    "```samtools index UHR_Rep3.bam```\n",
    "\n",
    "**Alternatively**, you can run the following command to index all .bam files in the directory. This way is easier than running multiple commands.\n",
    "\n",
    "```find *.bam -exec echo samtools index {} \\; | sh```\n",
    "\n",
    "Load both the UHR.bam and HBR.bam files in IGV. Make sure the index file .bam.bai is in the same directory."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e28dbc73-66c6-4175-858a-22952ae52d07",
   "metadata": {},
   "source": [
    "***\n",
    "# BAM read counting\n",
    "\n",
    "Make a new directory for the readcounts.\n",
    "\n",
    "```mkdir bam_readcount && cd bam_readcount```\n",
    "\n",
    "Create a faidx index for use with samtools mpileup. Place a copy of chr22_with_ERCC92.fa in the same directory below.\n",
    "\n",
    "```samtools faidx /home/user/bioinformatics/ref_genome/test_data/bam_readcount/chr22_with_ERCC92.fa```\n",
    "\n",
    "Run mpileup on a region of interest (ROI).\n",
    "\n",
    "```samtools mpileup -f /home/user/bioinformatics/ref_genome/test_data/bam_readcount/chr22_with_ERCC92.fa -r 22:18918457-18918467 /home/user/bioinformatics/ref_genome/test_data/align_output/UHR.bam /home/user/bioinformatics/ref_genome/test_data/align_output/HBR.bam```\n",
    "\n",
    "The output will only exist in the terminal. As per the tutorial, \"Each line consists of chromosome, 1-based coordinate, reference base, the number of reads covering the site, read bases and base qualities. At the read base column, a dot stands for a match to the reference base on the forward strand, a comma for a match on the reverse strand, ```ACGTN``` for a mismatch on the forward strand and ```acgtn``` for a mismatch on the reverse strand. A pattern ```\\+[0-9]+[ACGTNacgtn]+``` indicates there is an insertion between this reference position and the next reference position. The length of the insertion is given by the integer in the pattern, followed by the inserted sequence. [mpileup docs](http://samtools.sourceforge.net/mpileup.shtml).\n",
    "\n",
    "In the ```bam_readcount``` directory, create a bed file with the variant position on the desired chromosome.\n",
    "\n",
    "```echo \"22 38483683 38483683\" > snvs.bed```\n",
    "\n",
    "Now we can run ```bam-readcount``` on both bam files\n",
    "\n",
    "```./bam-readcount -l /home/user/bioinformatics/ref_genome/test_data/bam_readcount/snvs.bed -f /home/user/bioinformatics/ref_genome/test_data/bam_readcount/chr22_with_ERCC92.fa /home/user/bioinformatics/ref_genome/test_data/align_output/UHR.bam 2>/dev/null 1>/home/user/bioinformatics/ref_genome/test_data/bam_readcount/UHR_bam-readcounts.txt```\n",
    "\n",
    "```./bam-readcount -l /home/user/bioinformatics/ref_genome/test_data/bam_readcount/snvs.bed -f /home/user/bioinformatics/ref_genome/test_data/bam_readcount/chr22_with_ERCC92.fa /home/user/bioinformatics/ref_genome/test_data/align_output/HBR.bam 2>/dev/null 1>/home/user/bioinformatics/ref_genome/test_data/bam_readcount/HBR_bam-readcounts.txt```\n",
    "\n",
    "You can pipe the readcounts file to a perl script to parse the read counts for each base.\n",
    "\n",
    "```cat UHR_bam-readcounts.txt | perl -ne '@data=split(\"\\t\", $_); @Adata=split(\":\", $data[5]); @Cdata=split(\":\", $data[6]); @Gdata=split(\":\", $data[7]); @Tdata=split(\":\", $data[8]); print \"UHR Counts\\t$data[0]\\t$data[1]\\tA: $Adata[1]\\tC: $Cdata[1]\\tT: $Tdata[1]\\tG: $Gdata[1]\\n\";'```\n",
    "\n",
    "```cat HBR_bam-readcounts.txt | perl -ne '@data=split(\"\\t\", $_); @Adata=split(\":\", $data[5]); @Cdata=split(\":\", $data[6]); @Gdata=split(\":\", $data[7]); @Tdata=split(\":\", $data[8]); print \"HBR Counts\\t$data[0]\\t$data[1]\\tA: $Adata[1]\\tC: $Cdata[1]\\tT: $Tdata[1]\\tG: $Gdata[1]\\n\";'```\n",
    "\n",
    "Outputs:\n",
    "```UHR Counts      22      38483683        A: 163  C: 0    T: 0    G: 163```\n",
    "```HBR Counts      22      38483683        A: 75   C: 0    T: 0    G: 131```\n",
    "\n",
    "[Python tutorial for bam-readcount](https://github.com/genome/bam-readcount/tree/master/tutorial)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74f2d997-b225-44ff-b576-319544f4e913",
   "metadata": {},
   "source": [
    "***\n",
    "# Alignment QC\n",
    "\n",
    "Use samtools to view the format of the alignment bam files.\n",
    "\n",
    "```samtools view UHR.bam | head | column -t | less -S```\n",
    "```samtools view HBR.bam | head | column -t | less -S```\n",
    "\n",
    "You can add additional filtering flags to the samtools command. [This useful guide explains all of them](https://broadinstitute.github.io/picard/explain-flags.html). ```-f``` are required flags and ```-F``` are filtering flags.\n",
    "\n",
    "```samtools view -f 3 -F 268 UHR.bam | head | column -t | less -S```\n",
    "\n",
    "Here the ```-f 3``` requires the alignments are paired and 'mapped in a proper pair'. The ```-F 268``` means 'unmapped', 'mate is unmapped', and 'not primary alignment'.\n",
    "\n",
    "Look for alignments that are only for 'PCR or optical duplicates'.\n",
    "\n",
    "```samtools view -f 1024 UHR.bam | head```\n",
    "\n",
    "Output: nothing\n",
    "\n",
    "The command ```samtools flagstat``` can provide a summary of the alignments.\n",
    "\n",
    "```mkdir flagstat```\n",
    "\n",
    "```samtools flagstat HBR_Rep1.bam > flagstat/HBR_Rep1.bam.flagstat```\n",
    "\n",
    "```samtools flagstat HBR_Rep2.bam > flagstat/HBR_Rep2.bam.flagstat```\n",
    "\n",
    "```samtools flagstat HBR_Rep3.bam > flagstat/HBR_Rep3.bam.flagstat```\n",
    "\n",
    "```samtools flagstat UHR_Rep1.bam > flagstat/UHR_Rep1.bam.flagstat```\n",
    "\n",
    "```samtools flagstat UHR_Rep2.bam > flagstat/UHR_Rep2.bam.flagstat```\n",
    "\n",
    "```samtools flagstat UHR_Rep3.bam > flagstat/UHR_Rep3.bam.flagstat```\n",
    "\n",
    "In this instance, you can also run an expression to do it in one command.\n",
    "\n",
    "```find *Rep*.bam -exec echo samtools flagstat {} \\> flagstat/{}.flagstat \\; | sh```\n",
    "\n",
    "### FastQc on bam file\n",
    "\n",
    "From your alignment directory (where the bam files are).\n",
    "\n",
    "```mkidr fastqc```\n",
    "\n",
    "```fastqc UHR_Rep1.bam UHR_Rep2.bam UHR_Rep3.bam HBR_Rep1.bam HBR_Rep2.bam HBR_Rep3.bam```\n",
    "\n",
    "Clean up the current directory and move the reports into the fastqc folder.\n",
    "\n",
    "```mv *fastqc.html fastqc/```\n",
    "\n",
    "```mv *fastqc.zip fastqc/```\n",
    "\n",
    "### Picard\n",
    "\n",
    "Using Picard we can generate metrics and figures that express the quality of our RNA-seq data.\n",
    "\n",
    "We first need to create input files for Picard based on the reference genome so it can properly run the function CollectRnaSeqMetrics.\n",
    "\n",
    "```java -jar picard.jar CreateSequenceDictionary -R /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.fa -O /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.dict```\n",
    "\n",
    "The next line is copied directly from the source tutorial.\n",
    "\n",
    "\"Create a bed file of the location of ribosomal sequences in our reference (first extract from the gtf then convert to bed). Note that here we pull all the \"rrna\" transcripts from the GTF. This is a good strategy for the whole transcriptome ... but on chr22 there is very little \"rrna\" content, leading to 0 coverage for all samples, so we are also adding a single protein coding ribosomal gene \"RRP7A\" (normally we would not do this)\"\n",
    "\n",
    "```grep --color=none -i -P \"rrna|rrp7a\" /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.gtf > /home/user/bioinformatics/ref_genome/ref_ribosome.gtf```\n",
    "\n",
    "Note: ```gff2bed``` is part of BEDOPS. Run the script directly from the bedops/bin folder. You may need to add the /bin folder to your path variable ```export PATH=\"$PATH:/home/user/bioinformatics/bedops_linux_x86_64-v2.4.40/bin\"```, otherwise convert2bed may not be found.\n",
    "\n",
    "```gff2bed < /home/user/bioinformatics/ref_genome/ref_ribosome.gtf > /home/user/bioinformatics/ref_genome/ref_ribosome.bed```\n",
    "\n",
    "Create interval list file to identify the location of the ribosomal sequences in the reference\n",
    "\n",
    "```java -jar /home/user/bioinformatics/picard.jar BedToIntervalList -I /home/user/bioinformatics/ref_genome/ref_ribosome.bed -O /home/user/bioinformatics/ref_genome/ref_ribosome.interval_list -SD /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.dict```\n",
    "\n",
    "Create genePred file for the reference transcriptome\n",
    "\n",
    "```gtfToGenePred -genePredExt /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.gtf /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.ref_flat.txt```\n",
    "\n",
    "Reformat the genePred file\n",
    "\n",
    "```cat /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.ref_flat.txt | awk '{print $12\"\\t\"$0}' | cut -d$'\\t' -f1-11 > /home/user/bioinformatics/ref_genome/tmp.txt```\n",
    "\n",
    "```mv /home/user/bioinformatics/ref_genome/tmp.txt /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.ref_flat.txt```\n",
    "\n",
    "Go to the directory with your alignments.\n",
    "\n",
    "```mkdir picard```\n",
    "\n",
    "```find *Rep*.bam -exec echo java -jar /home/user/bioinformatics/picard.jar CollectRnaSeqMetrics I={} O=picard/{}.RNA_Metrics REF_FLAT=/home/user/bioinformatics/ref_genome/chr22_with_ERCC92.ref_flat.txt STRAND=SECOND_READ_TRANSCRIPTION_STRAND RIBOSOMAL_INTERVALS=/home/user/bioinformatics/ref_genome/ref_ribosome.interval_list \\; | sh```\n",
    "\n",
    "### RSeQC (optional)\n",
    "\n",
    "Another tool for generating RNA-seq QC reports.\n",
    "\n",
    "You will need\n",
    "- aligned bam files\n",
    "- index file for each bam file\n",
    "- transcript bed file (in bed12 format)\n",
    "\n",
    "Start by converting the reference gtf to genePred\n",
    "\n",
    "```gtfToGenePred /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.gtf /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.genePred```\n",
    "\n",
    "Then convert the genePred to bed12\n",
    "\n",
    "```genePredToBed /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.genePred /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.bed12```\n",
    "\n",
    "Go back to the alignment directory. Make sure the bioinformatics env is active.\n",
    "\n",
    "```cd /home/user/bioinformatics/ref_genome/test_data/align_output && mkdir rseqc```\n",
    "\n",
    "```geneBody_coverage.py -i UHR_Rep1.bam,UHR_Rep2.bam,UHR_Rep3.bam -r /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.bed12 -o rseqc/UHR```\n",
    "\n",
    "```geneBody_coverage.py -i HBR_Rep1.bam,HBR_Rep2.bam,HBR_Rep3.bam -r /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.bed12 -o rseqc/HBR```\n",
    "\n",
    "The following commands should also be run from the alignment directory ```cd /home/user/bioinformatics/ref_genome/test_data/align_output```\n",
    "\n",
    "```find *Rep*.bam -exec echo inner_distance.py -i {} -r /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.bed12 -o rseqc/{} \\; | sh```\n",
    "\n",
    "```find *Rep*.bam -exec echo junction_annotation.py -i {} -r /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.bed12 -o rseqc/{} \\; | sh```\n",
    "\n",
    "```find *Rep*.bam -exec echo junction_saturation.py -i {} -r /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.bed12 -o rseqc/{} \\; | sh```\n",
    "\n",
    "```find *Rep*.bam -exec echo read_distribution.py  -i {} -r /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.bed12 \\> rseqc/{}.read_dist.txt \\; | sh```\n",
    "\n",
    "```find *Rep*.bam -exec echo RNA_fragment_size.py -i {} -r /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.bed12 \\> rseqc/{}.frag_size.txt \\; | sh```\n",
    "\n",
    "```find *Rep*.bam -exec echo bam_stat.py -i {} \\> {}.bam_stat.txt \\; | sh```\n",
    "\n",
    "```rm -f log.txt```\n",
    "\n",
    "### MultiQC\n",
    "\n",
    "Now we will compile a multiQC report from all of the QC done above.\n",
    "\n",
    "In the alignment directory ```cd /home/user/bioinformatics/ref_genome/test_data/align_output```.\n",
    "\n",
    "```multiqc ./```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "763b3cf1-840a-4f47-950a-60d602ff597f",
   "metadata": {},
   "source": [
    "***\n",
    "# Expression Analysis\n",
    "\n",
    "Some notes before we start.\n",
    "\n",
    "First, there are multipls things you can do with your aligned reads from HISAT2. This pipeline is written with the intention of going through the differential expression analysis and visualizing it. The tutorial course can go even further than that, which I may or may not include in this. But, you can also just use **HTSEQ-count** to get raw read counts instead of FPKM values. \n",
    "\n",
    "FPKM = fragments per kilobase of transcript per million mapped reads. FPKM attempts to normalize for gene size and library depth. RPKM is essentially the same thing, however there are differing opinions on just how similar they are, and which is better. \n",
    "\n",
    "Mathematically, FPKM can be represented as:\n",
    "\n",
    "```FPKM = (C / (N / 1000000)) / (L / 1000)```, where C = number of mappable fragments for a gene, N = total number of mappable fragments in the library, and L = number of base pairs in the gene.\n",
    "\n",
    "Another normalization metric that may be mentioned is TPM (transcript per kilobase million). The only difference between FPKM and TPM is the order of operations.\n",
    "\n",
    "### Stringtie\n",
    "\n",
    "If you want to know more about the math behind how Stringtie works, you can [read the paper](https://www.nature.com/articles/nbt.3122).\n",
    "\n",
    "An outline of the Stringtie flags used:\n",
    "\n",
    "\n",
    "- ```–rf``` tells StringTie that our data is stranded and to use the correct strand specific mode (i.e. assume a stranded library fr-firststrand).\n",
    "- ```-p 8``` tells StringTie to use eight CPUs\n",
    "- ```-G``` reference annotation to use for guiding the assembly process (GTF/GFF3)\n",
    "- ```-e``` only estimate the abundance of given reference transcripts (requires -G)\n",
    "- ```-B``` enable output of Ballgown table files which will be created in the same directory as the output GTF (requires -G, -o recommended)\n",
    "- ```-o``` output path/file name for the assembled transcripts GTF (default: stdout)\n",
    "- ```-A``` output path/file name for gene abundance estimates\n",
    "\n",
    "I'll make this folder within the alignment folder ```cd /home/user/bioinformatics/ref_genome/test_data/align_output```.\n",
    "\n",
    "```mkdir -p expression/stringtie/ref_only/ && cd expression/stringtie/ref_only/```\n",
    "\n",
    "```./stringtie --rf -p 24 -G /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.gtf -e -B -o /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/HBR_Rep1/transcripts.gtf -A /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/HBR_Rep1/gene_abundances.tsv /home/user/bioinformatics/ref_genome/test_data/align_output/HBR_Rep1.bam```\n",
    "\n",
    "```./stringtie --rf -p 24 -G /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.gtf -e -B -o /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/HBR_Rep2/transcripts.gtf -A /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/HBR_Rep2/gene_abundances.tsv /home/user/bioinformatics/ref_genome/test_data/align_output/HBR_Rep2.bam```\n",
    "\n",
    "```./stringtie --rf -p 24 -G /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.gtf -e -B -o /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/HBR_Rep3/transcripts.gtf -A /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/HBR_Rep3/gene_abundances.tsv /home/user/bioinformatics/ref_genome/test_data/align_output/HBR_Rep3.bam```\n",
    "\n",
    "```./stringtie --rf -p 24 -G /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.gtf -e -B -o /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/UHR_Rep1/transcripts.gtf -A /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/UHR_Rep1/gene_abundances.tsv /home/user/bioinformatics/ref_genome/test_data/align_output/UHR_Rep1.bam```\n",
    "\n",
    "```./stringtie --rf -p 24 -G /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.gtf -e -B -o /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/UHR_Rep3/transcripts.gtf -A /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/UHR_Rep3/gene_abundances.tsv /home/user/bioinformatics/ref_genome/test_data/align_output/UHR_Rep3.bam```\n",
    "\n",
    "Have a look at what the raw output from StringTie looks like.\n",
    "\n",
    "```less -S /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/UHR_Rep1/transcripts.gtf```\n",
    "\n",
    "Improve formatting of the *less* output.\n",
    "\n",
    "```grep -v \"^#\" /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/UHR_Rep1/transcripts.gtf | grep -w \"transcript\" | column -t | less -S```\n",
    "\n",
    "You can also limit the view to just the transcript records and the correspoding FPKM values.\n",
    "\n",
    "```awk '{if ($3==\"transcript\") print}' /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/UHR_Rep1/transcripts.gtf | cut -f 1,4,9 | column -t | less -S```\n",
    "\n",
    "You can also view the transcript level expression values for each set of reads in the two files ```t_data.ctab``` and ```gene_abundances.tsv```. They should be located in the folders ```~/expression/stringtie/ref_only/``` and then in the UHR or HBR folders.\n",
    "\n",
    "Download the script for creating an expression matrix based off the Stringtie outputs.\n",
    "\n",
    "```cd /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only```\n",
    "\n",
    "```wget https://raw.githubusercontent.com/griffithlab/rnabio.org/master/assets/scripts/stringtie_expression_matrix.pl```\n",
    "\n",
    "```chmod +x stringtie_expression_matrix.pl```\n",
    "\n",
    "```./stringtie_expression_matrix.pl --expression_metric=TPM --result_dirs='HBR_Rep1,HBR_Rep2,HBR_Rep3,UHR_Rep1,UHR_Rep2,UHR_Rep3' --transcript_matrix_file=transcript_tpm_all_samples.tsv --gene_matrix_file=gene_tpm_all_samples.tsv```\n",
    "\n",
    "```./stringtie_expression_matrix.pl --expression_metric=FPKM --result_dirs='HBR_Rep1,HBR_Rep2,HBR_Rep3,UHR_Rep1,UHR_Rep2,UHR_Rep3' --transcript_matrix_file=transcript_fpkm_all_samples.tsv --gene_matrix_file=gene_fpkm_all_samples.tsv```\n",
    "\n",
    "```./stringtie_expression_matrix.pl --expression_metric=Coverage --result_dirs='HBR_Rep1,HBR_Rep2,HBR_Rep3,UHR_Rep1,UHR_Rep2,UHR_Rep3' --transcript_matrix_file=transcript_coverage_all_samples.tsv --gene_matrix_file=gene_coverage_all_samples.tsv```\n",
    "\n",
    "View the TPM output files.\n",
    "\n",
    "```column -t transcript_tpm_all_samples.tsv | less -S```\n",
    "\n",
    "```column -t gene_tpm_all_samples.tsv | less -S```\n",
    "\n",
    "### Generate raw counts instead of FPKM/TPM values\n",
    "\n",
    "If you only want to generate the raw counts this can be done using HTSeq-count [docs](https://htseq.readthedocs.io/).\n",
    "\n",
    "Some notes about the flags used:\n",
    "\n",
    "\n",
    "- ```–format``` specify the input file format one of BAM or SAM. Since we have BAM format files, select ‘bam’ for this option.\n",
    "- ```–order``` provide the expected sort order of the input file. Previously we generated position sorted BAM files so use ‘pos’.\n",
    "- ```–mode``` determines how to deal with reads that overlap more than one feature. We believe the ‘intersection-strict’ mode is best.\n",
    "- ```–stranded``` specifies whether data is stranded or not. The TruSeq strand-specific RNA libraries suggest the ‘reverse’ option for this parameter.\n",
    "- ```–minaqual``` will skip all reads with alignment quality lower than the given minimum value\n",
    "- ```–type``` specifies the feature type (3rd column in GFF file) to be used. (default, suitable for RNA-Seq and Ensembl GTF files: exon)\n",
    "- ```–idattr``` The feature ID used to identify the counts in the output table. The default, suitable for RNA-SEq and Ensembl GTF files, is gene_id.\n",
    "\n",
    "Change directories to whre the expression folder is ```cd /home/user/bioinformatics/ref_genome/test_data/align_output```\n",
    "\n",
    "```mkdir -p expression/htseq_counts && cd expression/htseq_counts```\n",
    "\n",
    "Run each htseq-count command to get the raw counts for each read file.\n",
    "\n",
    "```htseq-count --format bam --order pos --mode intersection-strict --stranded reverse --minaqual 1 --type exon --idattr gene_id /home/user/bioinformatics/ref_genome/test_data/align_output/UHR_Rep1.bam /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.gtf > UHR_Rep1_gene.tsv```\n",
    "\n",
    "```htseq-count --format bam --order pos --mode intersection-strict --stranded reverse --minaqual 1 --type exon --idattr gene_id /home/user/bioinformatics/ref_genome/test_data/align_output/UHR_Rep2.bam /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.gtf > UHR_Rep2_gene.tsv```\n",
    "\n",
    "```htseq-count --format bam --order pos --mode intersection-strict --stranded reverse --minaqual 1 --type exon --idattr gene_id /home/user/bioinformatics/ref_genome/test_data/align_output/UHR_Rep3.bam /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.gtf > UHR_Rep3_gene.tsv```\n",
    "\n",
    "```htseq-count --format bam --order pos --mode intersection-strict --stranded reverse --minaqual 1 --type exon --idattr gene_id /home/user/bioinformatics/ref_genome/test_data/align_output/HBR_Rep1.bam /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.gtf > HBR_Rep1_gene.tsv```\n",
    "\n",
    "```htseq-count --format bam --order pos --mode intersection-strict --stranded reverse --minaqual 1 --type exon --idattr gene_id /home/user/bioinformatics/ref_genome/test_data/align_output/HBR_Rep2.bam /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.gtf > HBR_Rep2_gene.tsv```\n",
    "\n",
    "```htseq-count --format bam --order pos --mode intersection-strict --stranded reverse --minaqual 1 --type exon --idattr gene_id /home/user/bioinformatics/ref_genome/test_data/align_output/HBR_Rep3.bam /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.gtf > HBR_Rep3_gene.tsv```\n",
    "\n",
    "I don't feel like paraphrasing this next section:\n",
    "\n",
    "\"Merge results files into a single matrix for use in edgeR. The following joins the results for each replicate together, adds a header, reformats the result as a tab delimited file, and shows you the first 10 lines of the resulting file :\"\n",
    "\n",
    "```cd /home/user/bioinformatics/ref_genome/test_data/align_output/expression/htseq_counts```\n",
    "\n",
    "```echo \"GeneID UHR_Rep1 UHR_Rep2 UHR_Rep3 HBR_Rep1 HBR_Rep2 HBR_Rep3\" > header.txtne.tsv | join - HBR_Rep2_gene.tsv | join - HBR_Rep3_gene.tsv > gene_read_counts_table_all.tsv```\n",
    "\n",
    "```echo \"GeneID UHR_Rep1 UHR_Rep2 UHR_Rep3 HBR_Rep1 HBR_Rep2 HBR_Rep3\" > header.txt```\n",
    "\n",
    "```cat header.txt gene_read_counts_table_all.tsv | grep -v \"__\" | awk -v OFS=\"\\t\" '$1=$1' > gene_read_counts_table_all_final.tsv```\n",
    "\n",
    "If you're curious, ```grep -v \"__\"``` is used to filter out summary lines at the end of htseq-count files that did not align, had poor quality alignment, etc.\n",
    "\n",
    "```rm -f gene_read_counts_table_all.tsv header.txt```\n",
    "\n",
    "```head gene_read_counts_table_all_final.tsv | column -t```\n",
    "\n",
    "### ERCC expression analysis\n",
    "\n",
    "I'm not sure how necessary this next part is at the moment, but since the files used in this version of the pipeline included ERCC spike-in controls I will go through it.\n",
    "\n",
    "```cd /home/user/bioinformatics/ref_genome/test_data/align_output/expression/htseq_counts```\n",
    "\n",
    "```wget http://genomedata.org/rnaseq-tutorial/ERCC_Controls_Analysis.txt```\n",
    "\n",
    "```cat ERCC_Controls_Analysis.txt```\n",
    "\n",
    "```wget https://github.com/griffithlab/rnabio.org/raw/master/assets/scripts/Tutorial_ERCC_expression.pl```\n",
    "\n",
    "```chmod +x Tutorial_ERCC_expression.pl```\n",
    "\n",
    "```./Tutorial_ERCC_expression.pl```\n",
    "\n",
    "```cat /home/user/bioinformatics/ref_genome/test_data/align_output/expression/htseq_counts/ercc_read_counts.tsv```\n",
    "\n",
    "```wget https://github.com/griffithlab/rnabio.org/raw/master/assets/scripts/Tutorial_ERCC_expression.R```\n",
    "\n",
    "```chmod +x Tutorial_ERCC_expression.R```\n",
    "\n",
    "```./Tutorial_ERCC_expression.R ercc_read_counts.tsv```\n",
    "\n",
    "Create plot for ERCC spike-in Stringtie TPM estimates vs. known concentration of ERCC spike-in mix.\n",
    "\n",
    "```cd /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only```\n",
    "\n",
    "```wget http://genomedata.org/rnaseq-tutorial/ERCC_Controls_Analysis.txt```\n",
    "\n",
    "```cat ERCC_Controls_Analysis.txt```\n",
    "\n",
    "```wget https://github.com/griffithlab/rnabio.org/raw/master/assets/scripts/Tutorial_ERCC_expression_tpm.pl```\n",
    "\n",
    "```chmod +x Tutorial_ERCC_expression_tpm.pl```\n",
    "\n",
    "```./Tutorial_ERCC_expression_tpm.pl```\n",
    "\n",
    "```cat /home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/ercc_tpm.tsv```\n",
    "\n",
    "```wget https://github.com/griffithlab/rnabio.org/raw/master/assets/scripts/Tutorial_ERCC_expression_tpm.R```\n",
    "\n",
    "```chmod +x Tutorial_ERCC_expression_tpm.R```\n",
    "\n",
    "```./Tutorial_ERCC_expression_tpm.R ercc_tpm.tsv```\n",
    "\n",
    "You should be able to view the generated figures under the ```~/htseq_counts``` and ```~/stringtie/ref_only``` directories. The figures will be in .pdf format."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9471a44c-1139-47a2-8c3b-561711ec3e05",
   "metadata": {},
   "source": [
    "***\n",
    "# Differential Expression\n",
    "\n",
    "The differential expression is done using the bioconductor package [Ballgown](https://www.bioconductor.org/packages/release/bioc/html/ballgown.html).\n",
    "\n",
    "```cd /home/user/bioinformatics/ref_genome/test_data/align_output```\n",
    "\n",
    "```mkdir -p de/ballgown/ref_only/ && cd de/ballgown/ref_only/```\n",
    "\n",
    "Start an R session by typing ```R``` in a terminal.\n",
    "\n",
    "The following code is an R script that will perform DE using Ballgown. You can run each command line by line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4955bb35-cc0f-4680-8c1b-8506204b948e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # load the required libraries\n",
    "# library(ballgown)\n",
    "# library(genefilter)\n",
    "# library(dplyr)\n",
    "# library(devtools)\n",
    "\n",
    "# # Create phenotype data needed for ballgown analysis\n",
    "# ids=c(\"UHR_Rep1\",\"UHR_Rep2\",\"UHR_Rep3\",\"HBR_Rep1\",\"HBR_Rep2\",\"HBR_Rep3\")\n",
    "# type=c(\"UHR\",\"UHR\",\"UHR\",\"HBR\",\"HBR\",\"HBR\")\n",
    "# results=\"/home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/\"\n",
    "# path=paste(results,ids,sep=\"\")\n",
    "# pheno_data=data.frame(ids,type,path)\n",
    "\n",
    "# # Load ballgown data structure and save it to a variable \"bg\"\n",
    "# bg = ballgown(samples=as.vector(pheno_data$path), pData=pheno_data)\n",
    "\n",
    "# # Display a description of this object\n",
    "# bg\n",
    "\n",
    "# # Load all attributes including gene name\n",
    "# bg_table = texpr(bg, 'all')\n",
    "# bg_gene_names = unique(bg_table[, 9:10])\n",
    "\n",
    "# # Save the ballgown object to a file for later use\n",
    "# save(bg, file='bg.rda')\n",
    "\n",
    "# # Perform differential expression (DE) analysis with no filtering\n",
    "# results_transcripts = stattest(bg, feature=\"transcript\", covariate=\"type\", getFC=TRUE, meas=\"FPKM\")\n",
    "# results_genes = stattest(bg, feature=\"gene\", covariate=\"type\", getFC=TRUE, meas=\"FPKM\")\n",
    "# results_genes = merge(results_genes, bg_gene_names, by.x=c(\"id\"), by.y=c(\"gene_id\"))\n",
    "\n",
    "# # Save a tab delimited file for both the transcript and gene results\n",
    "# write.table(results_transcripts, \"UHR_vs_HBR_transcript_results.tsv\", sep=\"\\t\", quote=FALSE, row.names = FALSE)\n",
    "# write.table(results_genes, \"UHR_vs_HBR_gene_results.tsv\", sep=\"\\t\", quote=FALSE, row.names = FALSE)\n",
    "\n",
    "# # Filter low-abundance genes. Here we remove all transcripts with a variance across the samples of less than one\n",
    "# bg_filt = subset (bg,\"rowVars(texpr(bg)) > 1\", genomesubset=TRUE)\n",
    "\n",
    "# # Load all attributes including gene name\n",
    "# bg_filt_table = texpr(bg_filt , 'all')\n",
    "# bg_filt_gene_names = unique(bg_filt_table[, 9:10])\n",
    "\n",
    "# # Perform DE analysis now using the filtered data\n",
    "# results_transcripts = stattest(bg_filt, feature=\"transcript\", covariate=\"type\", getFC=TRUE, meas=\"FPKM\")\n",
    "# results_genes = stattest(bg_filt, feature=\"gene\", covariate=\"type\", getFC=TRUE, meas=\"FPKM\")\n",
    "# results_genes = merge(results_genes, bg_filt_gene_names, by.x=c(\"id\"), by.y=c(\"gene_id\"))\n",
    "\n",
    "# # Output the filtered list of genes and transcripts and save to tab delimited files\n",
    "# write.table(results_transcripts, \"UHR_vs_HBR_transcript_results_filtered.tsv\", sep=\"\\t\", quote=FALSE, row.names = FALSE)\n",
    "# write.table(results_genes, \"UHR_vs_HBR_gene_results_filtered.tsv\", sep=\"\\t\", quote=FALSE, row.names = FALSE)\n",
    "\n",
    "# # Identify the significant genes with p-value < 0.05\n",
    "# sig_transcripts = subset(results_transcripts, results_transcripts$pval<0.05)\n",
    "# sig_genes = subset(results_genes, results_genes$pval<0.05)\n",
    "\n",
    "# # Output the significant gene results to a pair of tab delimited files\n",
    "# write.table(sig_transcripts, \"UHR_vs_HBR_transcript_results_sig.tsv\", sep=\"\\t\", quote=FALSE, row.names = FALSE)\n",
    "# write.table(sig_genes, \"UHR_vs_HBR_gene_results_sig.tsv\", sep=\"\\t\", quote=FALSE, row.names = FALSE)\n",
    "\n",
    "# # Exit the R session\n",
    "# quit(save=\"no\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ade1ba0a-314b-48a1-b492-3cec16eb4940",
   "metadata": {},
   "source": [
    "View the raw output from Ballgown.\n",
    "\n",
    "```head UHR_vs_HBR_gene_results.tsv | column -t```\n",
    "\n",
    "View how many genes are present on this chromosome.\n",
    "\n",
    "```grep -v feature UHR_vs_HBR_gene_results.tsv | wc -l```\n",
    "\n",
    "View the genes that passed the filtering step.\n",
    "\n",
    "```grep -v feature UHR_vs_HBR_gene_results_filtered.tsv | wc -l```\n",
    "\n",
    "View how many genes were considered differentially expressed (p < 0.05).\n",
    "\n",
    "```grep -v feature UHR_vs_HBR_gene_results_sig.tsv | wc -l```\n",
    "\n",
    "As an exercise or just a visualization, you can look at the top 20 DE genes in IGV and see if they seem to make sense. To explain the \n",
    "\n",
    "```grep -v feature UHR_vs_HBR_gene_results_sig.tsv | sort -rnk 3 | head -n 20 | column -t``` higher abundance in UHR\n",
    "\n",
    "```grep -v feature UHR_vs_HBR_gene_results_sig.tsv | sort -nk 3 | head -n 20 | column -t``` higher abundance in HBR\n",
    "\n",
    "Save all the genes that have a p < 0.05.\n",
    "\n",
    "```grep -v feature UHR_vs_HBR_gene_results_sig.tsv | cut -f 6 | sed 's/\\\"//g' > DE_genes.txt```\n",
    "\n",
    "```head DE_genes.txt```\n",
    "\n",
    "### ERCC DE analysis\n",
    "\n",
    "Also not sure if this section is necessary and how it will be done in other pipelines analyzing data using different spike-ins.\n",
    "\n",
    "```cd /home/user/bioinformatics/ref_genome/test_data/align_output/de/ballgown/ref_only```\n",
    "\n",
    "```wget https://raw.githubusercontent.com/griffithlab/rnabio.org/master/assets/scripts/Tutorial_ERCC_DE.R```\n",
    "\n",
    "```chmod +x Tutorial_ERCC_DE.R```\n",
    "\n",
    "```./Tutorial_ERCC_DE.R /home/user/bioinformatics/ref_genome/test_data/align_output/expression/htseq_counts/ERCC_Controls_Analysis.txt /home/user/bioinformatics/ref_genome/test_data/align_output/de/ballgown/ref_only/UHR_vs_HBR_gene_results.tsv```\n",
    "\n",
    "The results are output to a PDF file.\n",
    "\n",
    "***\n",
    "# edgeR analysis\n",
    "\n",
    "This is an alternative method of doing DE to the stringtie/ballgown method. This method uses the raw counts from htseq-count instead of the FPKM values.\n",
    "\n",
    "```cd /home/user/bioinformatics/ref_genome/test_data/align_output/de```\n",
    "\n",
    "```mkdir -p htseq_counts && cd htseq_counts```\n",
    "\n",
    "htseq-count only proivdes the Ensembl Gene ID for the counts and not the actual gene symbol. This next line attempts to do that using the GTF file for our transcriptome reference.\n",
    "\n",
    "```perl -ne 'if ($_ =~ /gene_id\\s\\\"(ENSG\\S+)\\\"\\;/) { $id = $1; $name = undef; if ($_ =~ /gene_name\\s\\\"(\\S+)\"\\;/) { $name = $1; }; }; if ($id && $name) {print \"$id\\t$name\\n\";} if ($_=~/gene_id\\s\\\"(ERCC\\S+)\\\"/){print \"$1\\t$1\\n\";}' /home/user/bioinformatics/ref_genome/chr22_with_ERCC92.gtf | sort | uniq > ENSG_ID2Name.txt```\n",
    "\n",
    "```head ENSG_ID2Name.txt```\n",
    "\n",
    "Count the unique gene ids.\n",
    "\n",
    "```cut -f 1 ENSG_ID2Name.txt | sort | uniq | wc -l```\n",
    "\n",
    "Count the unique gene names.\n",
    "\n",
    "```cut -f 2 ENSG_ID2Name.txt | sort | uniq | wc -l```\n",
    "\n",
    "Look at the most repeatd gene names.\n",
    "\n",
    "```cut -f 2 ENSG_ID2Name.txt | sort | uniq -c | sort -r | head```\n",
    "\n",
    "Run the R code in the cell below, uncomment everything first with ctrl + /."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "28ff1215-5574-47a6-a20c-fc6e0b5b620b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Set working directory where output will go\n",
    "# working_dir = \"/home/user/bioinformatics/ref_genome/test_data/align_output/de/htseq_counts\"\n",
    "# setwd(working_dir)\n",
    "\n",
    "# #Read in gene mapping\n",
    "# mapping=read.table(\"/home/user/bioinformatics/ref_genome/test_data/align_output/de/htseq_counts/ENSG_ID2Name.txt\", header=FALSE, stringsAsFactors=FALSE, row.names=1)\n",
    "\n",
    "# # Read in count matrix\n",
    "# rawdata=read.table(\"/home/user/bioinformatics/ref_genome/test_data/align_output/expression/htseq_counts/gene_read_counts_table_all_final.tsv\", header=TRUE, stringsAsFactors=FALSE, row.names=1)\n",
    "\n",
    "# # Check dimensions\n",
    "# dim(rawdata)\n",
    "\n",
    "# # Require at least 1/6 of samples to have expressed count >= 10\n",
    "# sample_cutoff <- (1/6)\n",
    "# count_cutoff <- 10\n",
    "\n",
    "# #Define a function to calculate the fraction of values expressed above the count cutoff\n",
    "# getFE <- function(data,count_cutoff){\n",
    "#  FE <- (sum(data >= count_cutoff)/length(data))\n",
    "#  return(FE)\n",
    "# }\n",
    "\n",
    "# #Apply the function to all genes, and filter out genes not meeting the sample cutoff\n",
    "# fraction_expressed <- apply(rawdata, 1, getFE, count_cutoff)\n",
    "# keep <- which(fraction_expressed >= sample_cutoff)\n",
    "# rawdata <- rawdata[keep,]\n",
    "\n",
    "# # Check dimensions again to see effect of filtering\n",
    "# dim(rawdata)\n",
    "\n",
    "# #################\n",
    "# # Running edgeR #\n",
    "# #################\n",
    "\n",
    "# # load edgeR\n",
    "# library('edgeR')\n",
    "\n",
    "# # make class labels\n",
    "# class <- c( rep(\"UHR\",3), rep(\"HBR\",3) )\n",
    "\n",
    "# # Get common gene names\n",
    "# Gene=rownames(rawdata)\n",
    "# Symbol=mapping[Gene,1]\n",
    "# gene_annotations=cbind(Gene,Symbol)\n",
    "\n",
    "# # Make DGEList object\n",
    "# y <- DGEList(counts=rawdata, genes=gene_annotations, group=class)\n",
    "# nrow(y)\n",
    "\n",
    "# # TMM Normalization\n",
    "# y <- calcNormFactors(y)\n",
    "\n",
    "# # Estimate dispersion\n",
    "# y <- estimateCommonDisp(y, verbose=TRUE)\n",
    "# y <- estimateTagwiseDisp(y)\n",
    "\n",
    "# # Differential expression test\n",
    "# et <- exactTest(y)\n",
    "\n",
    "# # Extract raw counts to add back onto DE results\n",
    "# counts <- getCounts(y)\n",
    "\n",
    "# # Print top genes\n",
    "# topTags(et)\n",
    "\n",
    "# # Print number of up/down significant genes at FDR = 0.05  significance level\n",
    "# summary(de <- decideTestsDGE(et, adjust.method=\"BH\", p=.05))\n",
    "\n",
    "# #Get output with BH-adjusted FDR values - all genes, any p-value, unsorted\n",
    "# out <- topTags(et, n = \"Inf\", adjust.method=\"BH\", sort.by=\"none\", p.value=1)$table\n",
    "\n",
    "# #Add raw counts back onto results for convenience (make sure sort and total number of elements allows proper join)\n",
    "# out2 <- cbind(out, counts)\n",
    "\n",
    "# #Limit to significantly DE genes\n",
    "# out3 <- out2[as.logical(de),]\n",
    "\n",
    "# # Order by p-value\n",
    "# o <- order(et$table$PValue[as.logical(de)],decreasing=FALSE)\n",
    "# out4 <- out3[o,]\n",
    "\n",
    "# # Save table\n",
    "# write.table(out4, file=\"DE_genes.txt\", quote=FALSE, row.names=FALSE, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de5046d5-8a4e-47b5-bc14-827a7ef1d122",
   "metadata": {},
   "source": [
    "***\n",
    "Compare the significantly DE genes between the 2 methods.\n",
    "\n",
    "```cat /home/user/bioinformatics/ref_genome/test_data/align_output/de/ballgown/ref_only/DE_genes.txt```\n",
    "\n",
    "```cat /home/user/bioinformatics/ref_genome/test_data/align_output/de/htseq_counts/DE_genes.txt```\n",
    "\n",
    "Pull out the gene IDs from both files.\n",
    "\n",
    "```cd /home/user/bioinformatics/ref_genome/test_data/align_output/de```\n",
    "\n",
    "```cut -f 1 /home/user/bioinformatics/ref_genome/test_data/align_output/de/ballgown/ref_only/DE_genes.txt | sort  > ballgown_DE_gene_symbols.txt```\n",
    "\n",
    "```cut -f 2 /home/user/bioinformatics/ref_genome/test_data/align_output/de/htseq_counts/DE_genes.txt | sort | uniq | grep -v Gene_Name > htseq_counts_edgeR_DE_gene_symbols.txt```\n",
    "\n",
    "You can use online tools to quickly create a venn diagram comparing the DE genes from the two methods. You can use the 2 cat commands, then copy/paste the outputs to the websites. Or just open the .txt files and copy from there. [biovenn.nl](https://www.biovenn.nl/) or [venny](https://bioinfogp.cnb.csic.es/tools/venny/).\n",
    "\n",
    "```cat /home/user/bioinformatics/ref_genome/test_data/align_output/de/ballgown_DE_gene_symbols.txt```\n",
    "\n",
    "```cat /home/user/bioinformatics/ref_genome/test_data/align_output/de/htseq_counts_edgeR_DE_gene_symbols.txt```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ab89416-0967-4b6e-a136-9cbcc4745913",
   "metadata": {},
   "source": [
    "***\n",
    "# Ballgown DE Visualization\n",
    "\n",
    "```cd /home/user/bioinformatics/ref_genome/test_data/align_output/de/ballgown/ref_only```\n",
    "\n",
    "Run the R code in the cell below. Uncomment all the lines with ctrl + / if needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "97302aa2-03a9-4482-8cef-74f009acff1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #load libraries\n",
    "# library(ballgown)\n",
    "# library(genefilter)\n",
    "# library(dplyr)\n",
    "# library(devtools)\n",
    "\n",
    "# #designate output file\n",
    "# outfile=\"/home/user/bioinformatics/ref_genome/test_data/align_output/de/ballgown/ref_only/Tutorial_Part2_ballgown_output.pdf\"\n",
    "\n",
    "# # Generate phenotype data\n",
    "# ids = c(\"UHR_Rep1\",\"UHR_Rep2\",\"UHR_Rep3\",\"HBR_Rep1\",\"HBR_Rep2\",\"HBR_Rep3\")\n",
    "# type = c(\"UHR\",\"UHR\",\"UHR\",\"HBR\",\"HBR\",\"HBR\")\n",
    "# results = \"/home/user/bioinformatics/ref_genome/test_data/align_output/expression/stringtie/ref_only/\"\n",
    "# path = paste(results,ids,sep=\"\")\n",
    "# pheno_data = data.frame(ids,type,path)\n",
    "\n",
    "# # Display the phenotype data\n",
    "# pheno_data\n",
    "\n",
    "# # Load the ballgown object from file\n",
    "# # Make sure to set the path to 'bg.rda', this file was created in the ballgown DE R script\n",
    "# load('/home/user/bioinformatics/ref_genome/test_data/align_output/de/ballgown/ref_only/bg.rda')\n",
    "\n",
    "# # The load command, loads an R object from a file into memory in our R session.\n",
    "# # You can use ls() to view the names of variables that have been loaded\n",
    "# ls()\n",
    "\n",
    "# # Print a summary of the ballgown object\n",
    "# bg\n",
    "\n",
    "# # Open a PDF file where we will save some plots. We will save all figures and then view the PDF at the end\n",
    "# pdf(file=outfile)\n",
    "\n",
    "# # Extract FPKM values from the 'bg' object\n",
    "# fpkm = texpr(bg,meas=\"FPKM\")\n",
    "\n",
    "# # View the last several rows of the FPKM table\n",
    "# tail(fpkm)\n",
    "\n",
    "# # Transform the FPKM values by adding 1 and convert to a log2 scale\n",
    "# fpkm = log2(fpkm+1)\n",
    "\n",
    "# # View the last several rows of the transformed FPKM table\n",
    "# tail(fpkm)\n",
    "\n",
    "# # Create boxplots to display summary statistics for the FPKM values for each sample\n",
    "# boxplot(fpkm,col=as.numeric(as.factor(pheno_data$type))+1,las=2,ylab='log2(FPKM+1)')\n",
    "\n",
    "# # col=as.numeric(as.factor(pheno_data$type))+1 - set color based on pheno_data$type which is UHR vs. HBR\n",
    "# # las=2 - labels are perpendicular to axis \n",
    "# # ylab='log2(FPKM+1)' - set ylab to indicate that values are log2 transformed\n",
    "\n",
    "# # Display the transcript ID for a single row of data\n",
    "# ballgown::transcriptNames(bg)[2763]\n",
    "\n",
    "# # Display the gene name for a single row of data\n",
    "# ballgown::geneNames(bg)[2763]\n",
    "\n",
    "# # Create a BoxPlot comparing the expression of a single gene for all replicates of both conditions\n",
    "# boxplot(fpkm[2763,] ~ pheno_data$type, border=c(2,3), main=paste(ballgown::geneNames(bg)[2763],' : ', ballgown::transcriptNames(bg)[2763]),pch=19, xlab=\"Type\", ylab='log2(FPKM+1)')\n",
    "\n",
    "# # border=c(2,3) - set border color for each of the boxplots\n",
    "# # main=paste(ballgown::geneNames(bg)[2763],' : ', ballgown::transcriptNames(bg)[2763]) - set title to gene : transcript\n",
    "# # xlab=\"Type\" - set x label to Type\n",
    "# # ylab='log2(FPKM+1)' - set ylab to indicate that values are log2 transformed\n",
    "\n",
    "\n",
    "# # Add the FPKM values for each sample onto the plot\n",
    "# points(fpkm[2763,] ~ jitter(c(2,2,2,1,1,1)), col=c(2,2,2,1,1,1)+1, pch=16)\n",
    "# # pch=16 - set plot symbol to solid circle, default is empty circle\n",
    "\n",
    "\n",
    "# # Create a plot of transcript structures observed in each replicate and color transcripts by expression level\n",
    "# plotTranscripts(ballgown::geneIDs(bg)[2763], bg, main=c('TST in all HBR samples'), sample=c('HBR_Rep1', 'HBR_Rep2', 'HBR_Rep3'), labelTranscripts=TRUE)\n",
    "# plotTranscripts(ballgown::geneIDs(bg)[2763], bg, main=c('TST in all UHR samples'), sample=c('UHR_Rep1', 'UHR_Rep2', 'UHR_Rep3'), labelTranscripts=TRUE)\n",
    "\n",
    "# #plotMeans('TST',bg,groupvar=\"type\",legend=FALSE)\n",
    "\n",
    "# # Close the PDF device where we have been saving our plots\n",
    "# dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1447a81d-69d9-4c38-a758-a43eb7fa4f77",
   "metadata": {},
   "source": [
    "***\n",
    "# Batch Correction\n",
    "\n",
    "The dataset used for everything leading up to this step was done at a single center, at the same time, using the same methodology. Therefore batch effects are not expected. A different dataset will be used to demonstrate this process.\n",
    "\n",
    "It is highly recommended to read the entire [ComBat-seq paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7518324/) to better understand the concept of batch correction, and how it compares to normalization. **Cite the paper if this methodology is used**.\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
